---
title: "A Tidy Transcriptomics introduction to RNA-Seq analyses"
author:
  - Maria Doyle, Peter MacCallum Cancer Centre^[maria.doyle@petermac.org]
  - Stefano Mangiola, Walter and Eliza Hall Institute^[mangiola.s@wehi.edu.au]
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{TidyTranscriptomics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Workshop Description

This workshop will present how to perform analysis of RNA sequencing data following the tidy data paradigm. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions. 

This can be achieved for RNA sequencing data with the [tidybulk](github.com/stemangiola/tidybulk), [tidyHeatmap](github.com/stemangiola/tidyHeatmap) and tidyverse packages. The package [tidybulk](github.com/stemangiola/tidybulk) provides a tidy data structure and a modular framework for bulk transcriptional analyses. tidyHeatmap provides a tidy implementation of ComplexHeatmap. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data.

The topics presented in this workshop will be

- Data exploration
- Data dimensionality reduction and clustering
- Differential gene expression analysis 
- Data visualisation

## Pre-requisites

* Basic knowledge of RStudio
* Familiarity with tidyverse syntax

Recommended Background Reading 
[Introduction to R for Biologists](https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html)

## Workshop Participation

Students will be expected to participate in the workshop in a hands-on way, following along with the code provided and performing exercises.

## _R_ / _Bioconductor_ packages used

* tidyverse
* tidybulk
* tidyHeatmap
* edgeR
* ggrepel
* airway

## Time outline

| Activity                                     | Time |
|----------------------------------------------|------|
| Data preprocessing                           | 15m  |
| Data dimensionality reduction and clustering | 15m  |
| Differential gene expression                 | 10m  |
| Data visualisation                           | 20m  |

# Workshop goals and objectives

In exploring and analysing RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualisation, that need to be understood. These concepts can be intuitively explained to new users, however, (i) the use of a heterogeneous vocabulary and jargon by methodologies/algorithms/packages, (ii) the complexity of data wrangling, and (iii) the coding burden, impede effective learning of the statistics and biology underlying an informed RNA sequencing analysis. 

The tidytranscriptomics approach to RNA sequencing data analysis abstracts out the coding-related complexity and provides tools that use an intuitive and jargon-free vocabulary, enabling focus on the statistical and biological challenges.

## Learning goals

* To understand the key concepts and steps of bulk RNA sequencing data analysis
* To approach data representation and analysis though a tidy data paradigm, integrating tidyverse with tidybulk and tidyHeatmap.

## Learning objectives

* Recall the key concepts of RNA sequencing data analysis
* Apply the concepts to publicly available data
* Create plots that summarise the information content of the data and analysis results

## Acknowledgements
This material is adapted from an R for RNA-Seq workshop first run [here](http://combine-australia.github.io/2016-05-11-RNAseq/).

```{r, out.width = "200px"}
knitr::include_graphics(system.file(package='tidytranscriptomics',  'vignettes', 'tidybulk_logo.png'))
```

# Overview

* Setting up the data 
* Filtering lowly expressed genes
* Scaling of counts
* Exploratory analyses
* Differential expression analysis
* Visualisation (MA plot, Volcano plot)


# Introduction
Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA-Seq to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.

There are many steps involved in analysing an RNA-Seq experiment. Analysing an RNAseq experiment begins with sequencing reads. These are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and getting stuck into analysis.

First, letâ€™s load all the packages we will need to analyse the data.

```{r message=FALSE, warning=FALSE}
# load libraries

# tidyverse core packages
library(tibble)
library(dplyr)
library(tidyr)
library(readr)
library(stringr)
library(ggplot2)

# tidyverse-friendly packages
library(tidyHeatmap)
library(tidybulk)
library(ggrepel)
library(plotly)

# dataset
library(airway)
```

## Airway RNA-seq dataset
Here we will perform RNA-Seq analysis using the data from the *airway* package. It has 8 samples, a sample treated with dex and an untreated control sample for 4 cell lines.

## Setting up the data

The airway RNA-Seq data is stored as a *RangedSummarizedExperiment* object. We'll convert the SummarizedExperiment object into a tidybulk tibble, which is the tidyverse table format.

In this workshop we will be using the tidyverse pipe `%>%`. This 'pipes' the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps (as we'll see later). It also can make the steps clearer and easier to see. For more details on the pipe see [here](https://r4ds.had.co.nz/pipes.html).

```{r}
# load airway RNA-Seq data
data(airway)

# convert to tidybulk format
counts_airway <-
	airway %>%
	tidybulk()
```

You can type the name of the object to view the first few lines and to see how many rows and columns it has. 

```{r}
counts_airway
```

The `counts_tt` object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance expressed in number of reads aligning to the gene in each experimental sample. The remaining columns include sample-wise information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.

We can shorten the sample names. We can remove the SRR1039 prefix that's present in all of them, as shorter names can fit better in some of the plots we will create. We can use `mutate()` together with `str_replace()` to remove the SRR1039 string from the sample column.

```{r}
counts_format <- counts_airway %>% 
  mutate(sample=str_remove(sample, "SRR1039"))
```


## Adding gene symbols

We can get the gene symbols for these Ensembl gene ids with tidybulk's `ensembl_to_symbol`. This works for human and mouse.

```{r}
counts_tt <- ensembl_to_symbol(counts_format, feature)
counts_tt
```

With tidyverse, all above operations can be linked with the `%>%`, as shown below. This has the benefits that

* no temporary variables need to be created
* less typing is required
* the steps can be seen more clearly. 

```{r}
counts_tt <- 	
	airway %>%
  tidybulk() %>%
  mutate(sample=str_remove(sample, "SRR1039")) %>%
  ensembl_to_symbol(feature)
```

From this tidybulk tibble, we can perform differential expression analysis with the tidybulk package.

# Filtering lowly expressed genes  
Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.

With tidybulk, it is not really necessary to explicitly filter lowly transcribed genes, as all calculations (e.g., scaling, removal of unwanted variation, differential expression testing) are performed on abundantly transcribed genes; although in case of scaling, the scaling is applied back to all genes in the dataset. Tidybulk can use the *edgeR* `filterByExpr` function described [here](https://f1000research.com/articles/5-1408). By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is 2 samples. tidybulk performs this filtering in the functions we will use `scale_abundance` and `test_differential_abundance`.

# Scaling counts to normalise

Scaling of counts, normalisation, is performed to eliminate uninteresting differences between samples due to sequencing depth or composition. In the tidybulk package the function `scale_abundance` generates scaled counts, with scaling factors calculated on abundant (filtered) transcript and applied to all transcripts. We can choose from different normalisation methods. Here we will use the default, TMM, [@robinson2010tmm]. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.

```{r}
# Scaling counts
counts_scaled <- counts_tt %>% scale_abundance(factor_of_interest = dex)

# take a look
counts_scaled
```

After we run `scale_abundance` we should see some columns have been added at the end. We have a column called `lowly_abundant` that indicates whether the gene has been filtered due to being lowly expressed. FALSE means the gene was not filtered, TRUE means it was. The `counts_scaled` column contains the scaled counts.

We can now see the difference of abundance densities before and after scaling.

```{r }
counts_scaled %>%
	pivot_longer(cols = c("counts", "counts_scaled"), names_to = "source", values_to = "abundance") %>%
  filter(!lowly_abundant) %>%
  ggplot(aes(x=abundance + 1, group=sample, color=sample)) +
	geom_density() +
	facet_wrap(~source) +
	scale_x_log10() +
	theme_bw()
```

#### Exercises
1. To see the effect of the filtering, adapt the code above to create density plots of the genes without filtering lowly expressed. How does the with and without filtering compare?
2. Another way to check the distributions of the counts in the samples is with box plots. Adapt the code above to create box plots of the counts before and after scaling. How does the before and after compare?


As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from the paper [here](https://f1000research.com/articles/5-1408)).

**tidybulk**

```{r eval=FALSE}
# tidybulk
airway %>%
  tidybulk() %>%
  scale_abundance() %>%
	pivot_longer(cols = c("counts", "counts_scaled"), names_to = "source", values_to = "abundance") %>%
  filter(!lowly_abundant) %>%
  ggplot(aes(x=abundance + 1, group=sample, color=sample)) +
	geom_density() +
	facet_wrap(~source) +
	scale_x_log10() +
	theme_bw()
```


**base R using edgeR**

```{r eval=FALSE}
dgList <- SE2DGEList(airway)
keep.exprs <- filterByExpr(dgList, group=dgList$samples$dex)
dgList <- dgList[keep.exprs,, keep.lib.sizes=FALSE]
nsamples <- ncol(dgList)
col <- RColorBrewer::brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
logcounts <- log2(dgList$counts)
plot(density(logcounts[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title(main="Counts")
for (i in 2:nsamples){
  den <- density(logcounts[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", legend=dgList$samples$Run, text.col=col, bty="n")
dgList_norm <- calcNormFactors(dgList)
lcpm_n <- cpm(dgList_norm, log=TRUE)
plot(density(lcpm_n[,1]), col=col[1], lwd=2, ylim=c(0,0.26), las=2, main="", xlab="")
title("Counts scaled")
for (i in 2:nsamples){
  den <- density(lcpm_n[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
legend("topright", legend=dgList_norm$samples$Run, text.col=col, bty="n")
```


# Exploratory analyses

## Multidimensional scaling plots

By far, one of the most important plots we make when we analyse RNA-Seq data are MDS plots. An MDS plot is a visualisation of a principal components analysis, which determines the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don't need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the `reduce_dimensions` function to calculate the dimensions.


```{r}
# get MDS dimensions
counts_scal_MDS <-
  counts_scaled %>%
  reduce_dimensions(method="MDS", .dims = 2)
```

This joins the result to the counts object. For plotting, we can select just the dimensions for the samples with `pivot_sample`.

```{r}
# take a look
counts_scal_MDS %>% pivot_sample()
```

We can now plot the reduced dimensions

```{r}
# MDS plot
counts_scal_MDS %>%
	pivot_sample() %>%
	ggplot(aes(x=Dim1, y=Dim2, colour=dex)) + 
  geom_point() +
	theme_bw()
```


#### Exercises 

1. Colour the MDS plot by cell line. Try using `shape` inside `aes` with the dex or cell variables.
2. Discuss what is the greatest source of variation in the data (i.e. what does dimension 1 represent)?


## Hierarchical clustering with heatmaps

An alternative to MDS for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, `keep_variable` to extract the most variable genes which we can then plot with tidyHeatmap.

```{r fig.height=5, fig.width=8}
counts_scaled %>% 
	
	# extract 500 most variable genes
	keep_variable( .abundance = counts_scaled, top = 500) %>%
	
	# create heatmap
	heatmap(
	      .column = sample,
	      .row = feature,
	      .value = counts_scaled,
	      annotation = c(dex, cell),
	      transform = log1p 
	  )
```

# Differential expression

Now that we are happy that the data looks good, we can continue to testing for differentially expressed genes. We will use the `test_differential_abundance` from tidybulk which currently uses *edgeR* to perform the differential expression analysis. We give `test_differential_abundance` our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line we could use the formula `0 + dex`, however, each treated and untreated sample is from a different cell line so we add the cell line as an additional factor `0 + dex + cell`. We also provide the names of the groups we want to compare to .contrasts (e.g. .contrasts = c("dextreat - dexuntreat")). `test_differential_abundance` will perform the filtering of lowly expressed genes as described before. The results will be joined to our counts for every sample. If we just want a table of differentially expressed genes we can use `pivot_transcript`.

```{r warning=FALSE}
counts_de <- counts_tt %>%
    test_differential_abundance(
      .formula = ~ 0 + dex + cell,
      .contrasts = c("dextrt - dexuntrt"))

# take a look
counts_de %>%
	pivot_transcript()
```
Now we have columns with our logFC and FDR P values. There is also a column called significant that indicates if the gene has FDR < 0.05.

Tidybulk enables a simplified way of performing an RNA-seq differential expression testing (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.

**tidybulk using edgeR**
```{r eval=FALSE}
counts_m %>%
  tidybulk() %>%
  test_differential_abundance(.formula = ~ 0 + dex + cell, .contrasts = c("dextrt - dexuntrt"))
```

**standard edgeR**
```{r eval=FALSE}
dgList <- DGEList(counts=counts_m, group=dex)
keep <- filterByExpr(dgList)
dgList <- dgList[keep,,keep.lib.sizes=FALSE]
dgList <- calcNormFactors(dgList)
design <- model.matrix(~ 0 + dex + cell)
dgList <- estimateDisp(dgList,design)
fit <- glmQLFit(dgList,design)
qlf <- glmQLFTest(fit,coef=2)
```


### Table of differentially expressed genes

We can write out our differentially expressed genes to a file that can be loaded into e.g. Excel. `write_tsv` will create a tab-separated file.

```{r eval=FALSE}
# save results
counts_de %>% 
	pivot_transcript() %>%
	write_tsv("de_results.tsv")
```


### Counting differentially expressed genes

We only have one contrast here so we could remove the suffix `_dextrt - dexuntrt` from the column headers to make the names shorter to use in the rest of the commands here, using `rename_at` package together with `str_replace`.

```{r}
counts_de_pretty <- 
	counts_de %>%
  rename_at(vars(contains("dex")),
            ~str_replace(., "_dextrt - dexuntrt", ""))
```


We can count how many differentially expressed genes there are using the significant column.

```{r}
counts_de_pretty %>% 
  filter(significant == TRUE) %>% 
  summarise(num_de = n_distinct(transcript))
```

#### Exercise
Count how many upregulated DE genes there are.  
Count how many downregulated DE genes there are.

### Extracting top differentially expressed genes

We can see the top genes by smallest P value. We'll take a look at the top 6.
```{r}
topgenes <- counts_de_pretty %>%
	pivot_transcript() %>%
  arrange(PValue) %>%
  head(6)

topgenes
```	

We can extract the symbols for these top genes to use in some of the plots we will make.
```{r}
topgenes_symbols <- topgenes %>% pull(transcript)

# take a look
topgenes_symbols
```
	
# Plots after testing for DE

### Volcano plots

Volcano plots are a useful genome-wide plot for checking that the analysis looks good. Volcano plots enable us to visualise significance of expression (P value) versus logFC. Highly significant genes are towards the top of the plot. We can also colour significant genes (e.g. genes with FDR < 0.05) 

```{r}
# volcanoplot, minimal
counts_de_pretty %>%
  filter(!lowly_abundant) %>%
  ggplot(aes(x=logFC, y=-log10(PValue), colour=significant)) +
  geom_point() +
	theme_bw()
```

A more informative plot, integrating some of the packages in tidyverse.

```{r warning=FALSE}
counts_de_pretty %>%
    pivot_transcript() %>%

    # Subset data
		filter(!lowly_abundant) %>%
    mutate(significant = FDR<0.05 & abs(logFC) >=2) %>%
    mutate(transcript = ifelse(transcript %in% topgenes_symbols, as.character(transcript), "")) %>%

    # Plot
    ggplot(aes(x = logFC, y = -log10(PValue), label=transcript)) +
    geom_point(aes(color = significant, size = significant, alpha=significant)) +
    geom_text_repel() +
    scale_color_manual(values=c("black", "#e11f28")) +
    scale_size_discrete(range = c(0, 2)) +
    xlim(c(-10,10)) +
		theme_bw()
```

### Stripcharts

Before following up on the DE genes with further lab work, it is also recommended to have a look at the expression levels of the individual samples for the genes of interest. We can use stripcharts to do this. These will help show if expression is consistent amongst replicates in the groups.

With stripcharts we can see if replicates tend to group together and how the expression compares to the other groups. We'll also add a box plot to show the distribution.


```{r}
strip_chart = 
	counts_scaled %>%
	
	# extract counts for top DE genes
	filter(transcript %in% topgenes_symbols) %>%
	
	# make stripchart
	ggplot(aes(x = dex, y = counts_scaled + 1, fill = dex, label = sample)) +
	geom_boxplot() +
	geom_jitter() +
	facet_wrap(~transcript) +
	scale_y_log10()+
	theme_bw()

strip_chart
```

# Interactive Plots

A really nice feature of using tidyverse and ggplot2 is that we can make interactive plots quite easily using the plotly package. This can be very useful for exploring what genes or samples are in the plots. We can make interactive plots directly from our ggplot2 object (strip_chart). Having `label` into the `aes` is useful to visualise the identified of the data point or other variables to be visible when we hover over the plot.

We can also specify which parameters from the `aes` we want to show up when we hover over the plot with `tooltip`.

```{r, warning=FALSE}
strip_chart %>% ggplotly(tooltip = c("label", "y"))
```

# Key Points
- RNA-Seq data can be analysed in a 'tidy' way using the packages tidyverse, tidybulk and tidyHeatmap
- Key steps in an RNA-Seq analysis are filtering lowly expressed genes, normalisation for sequencing depth and composition, and testing for differential expression
- MDS plots are very important for examining the quality of the data
- Other useful plots for assessing RNA-Seq data are density plots, box plots, MA plots, volcano plots, stripcharts and heatmaps. These can all be made in a 'tidy' way.


# Further Reading
[RNA-Seq analysis is easy as 1-2-3 with limma, Glimma and edgeR](https://f1000research.com/articles/5-1408)  

# Reproducibility 
Record package and version information with `sessionInfo`

```{r}
sessionInfo()
```
