<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/bioc_2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/bioc_2020_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<div id="a-tidy-transcriptomics-introduction-to-rna-seq-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#a-tidy-transcriptomics-introduction-to-rna-seq-analyses" class="anchor"></a>A Tidy Transcriptomics introduction to RNA-Seq analyses</h1>
<p>Authors: Dr. Maria Doyle<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Dr. Stefano Mangiola<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p>
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>This workshop will present how to perform analysis of RNA sequencing data following the tidy data paradigm. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.</p>
<p>This can be achieved for RNA sequencing data with the <a href="github.com/stemangiola/tidybulk">tidybulk</a>, <a href="github.com/stemangiola/tidyHeatmap">tidyHeatmap</a> and tidyverse packages. The package <a href="github.com/stemangiola/tidybulk">tidybulk</a> provides a tidy data structure and a modular framework for bulk transcriptional analyses. tidyHeatmap provides a tidy implementation of ComplexHeatmap. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data.</p>
<p>The topics presented in this workshop will be</p>
<ul>
<li>Data exploration</li>
<li>Data dimensionality reduction and clustering</li>
<li>Differential gene expression analysis</li>
<li>Data visualisation</li>
</ul>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of RStudio</li>
<li>Familiarity with tidyverse syntax</li>
</ul>
<p>Recommended Background Reading <a href="https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html">Introduction to R for Biologists</a></p>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Students will be expected to participate in the workshop in a hands-on way, following along with the code provided and performing exercises.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li>tidyverse</li>
<li>tidybulk</li>
<li>tidyHeatmap</li>
<li>edgeR</li>
<li>ggrepel</li>
<li>airway</li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data exploration</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Data dimensionality reduction and clustering</td>
<td>30m</td>
</tr>
<tr class="odd">
<td>Differential gene expression</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Data visualisation</td>
<td>30m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="workshop-goals-and-objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h1>
<p>In exploring and analysing RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualisation, that need to be understood. These concepts can be intuitively explained to new users, however, (i) the use of a heterogeneous vocabulary and jargon by methodologies/algorithms/packages, (ii) the complexity of data wrangling, and (iii) the coding burden, impede effective learning of the statistics and biology underlying an informed RNA sequencing analysis.</p>
<p>The tidytranscriptomics approach to RNA sequencing data analysis abstracts out the coding-related complexity and provides tools that use an intuitive and jargon-free vocabulary, enabling focus on the statistical and biological challenges.</p>
<div id="learning-goals" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h2>
<ul>
<li>To understand the key concepts and steps of bulk RNA sequencing data analysis</li>
<li>To approach data representation and analysis though a tidy data paradigm, integrating tidyverse with tidybulk and tidyHeatmap.</li>
</ul>
</div>
<div id="learning-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h2>
<ul>
<li>Recall the key concepts of RNA sequencing data analysis</li>
<li>Apply the concepts to publicly available data</li>
<li>Create plots that summarise the information content of the data and analysis results</li>
</ul>
</div>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
<p>This material is adapted from an R for RNA-Seq workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>.</p>
<p><img src="../../../../tmp/Rtmp3mR8kt/temp_libpath1eca3c2bbe2a/tidytranscriptomics/vignettes/tidybulk_logo.png" width="501"></p>
</div>
</div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<ul>
<li>Setting up tables of counts and metadata</li>
<li>Formatting the data</li>
<li>Filtering lowly expressed genes</li>
<li>Scaling of counts</li>
<li>Quality control plots</li>
<li>Differential expression analysis</li>
<li>Visualisation (Heatmaps, MA plot, Volcano plot)</li>
</ul>
</div>
<div id="introduction-and-data-import" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-and-data-import" class="anchor"></a>Introduction and data import</h1>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA-Seq to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA-Seq experiment. Analysing an RNAseq experiment begins with sequencing reads. These are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and getting stuck into analysis.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># load libraries</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">utils</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">stats</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tibble</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">readr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">stringr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggrepel</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyHeatmap</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidybulk</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">airway</span>)</pre></body></html></div>
<div id="airway-rna-seq-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#airway-rna-seq-dataset" class="anchor"></a>Airway RNA-seq dataset</h2>
<p>Here we will perform RNA-Seq analysis using the data from airway package. It has 8 samples, a sample treated with dex and an untreated control sample for 4 cell lines.</p>
</div>
<div id="setting-up-the-tables" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-tables" class="anchor"></a>Setting up the tables</h2>
<p>The airway RNA-Seq data is stored as a SummarizedExperiment object. We’ll extract the counts into a table and do the same for the metadata information. Tables are often the format of the data at the start of an RNA-seq analysis.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">airway</span>)

<span class="co"># extract counts </span>
<span class="no">counts</span> <span class="kw">&lt;-</span> <span class="fu">assay</span>(<span class="no">airway</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="kw">rownames</span> <span class="kw">=</span> <span class="st">"GeneID"</span>)</pre></body></html></div>
<p>You can type the name of the object to view the first few lines and to see how many rows and columns it has.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">counts</span>
<span class="co">#&gt; # A tibble: 64,102 x 9</span>
<span class="co">#&gt;    GeneID SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516 SRR1039517</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;      &lt;int&gt;</span>
<span class="co">#&gt;  1 ENSG0…        679        448        873        408       1138       1047</span>
<span class="co">#&gt;  2 ENSG0…          0          0          0          0          0          0</span>
<span class="co">#&gt;  3 ENSG0…        467        515        621        365        587        799</span>
<span class="co">#&gt;  4 ENSG0…        260        211        263        164        245        331</span>
<span class="co">#&gt;  5 ENSG0…         60         55         40         35         78         63</span>
<span class="co">#&gt;  6 ENSG0…          0          0          2          0          1          0</span>
<span class="co">#&gt;  7 ENSG0…       3251       3679       6177       4252       6721      11027</span>
<span class="co">#&gt;  8 ENSG0…       1433       1062       1733        881       1424       1439</span>
<span class="co">#&gt;  9 ENSG0…        519        380        595        493        820        714</span>
<span class="co">#&gt; 10 ENSG0…        394        236        464        175        658        584</span>
<span class="co">#&gt; # … with 64,092 more rows, and 2 more variables: SRR1039520 &lt;int&gt;,</span>
<span class="co">#&gt; #   SRR1039521 &lt;int&gt;</span></pre></body></html></div>
<p>The <code>counts</code> object contains information about genes (one gene per row), the first column has the Ensembl gene id, and the remaining columns contain information about the number of reads aligning to the gene in each experimental sample.</p>
<p>Next we’ll extract the sample information, so we know what groups (treatment and cell line) the samples belong to.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">sampleinfo</span> <span class="kw">&lt;-</span> <span class="fu">colData</span>(<span class="no">airway</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="kw">rownames</span> <span class="kw">=</span> <span class="st">"sample"</span>)

<span class="no">sampleinfo</span>
<span class="co">#&gt; # A tibble: 8 x 10</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 SRR10… GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 SRR10… GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 SRR10… GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 SRR10… GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 SRR10… GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 SRR10… GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 SRR10… GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 SRR10… GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 1 more variable: BioSample &lt;fct&gt;</span></pre></body></html></div>
<p>The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.</p>
<p>First we will convert the counts into long format (tidy format), similar to what we did in the Intro to R session.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># convert to tidy format</span>
<span class="no">counts_long</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="no">counts</span>, <span class="kw">cols</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">"SRR"</span>), <span class="kw">names_to</span> <span class="kw">=</span> <span class="st">"sample"</span>, <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"count"</span>)

<span class="co"># take a look</span>
<span class="no">counts_long</span>
<span class="co">#&gt; # A tibble: 512,816 x 3</span>
<span class="co">#&gt;    GeneID          sample     count</span>
<span class="co">#&gt;    &lt;chr&gt;           &lt;chr&gt;      &lt;int&gt;</span>
<span class="co">#&gt;  1 ENSG00000000003 SRR1039508   679</span>
<span class="co">#&gt;  2 ENSG00000000003 SRR1039509   448</span>
<span class="co">#&gt;  3 ENSG00000000003 SRR1039512   873</span>
<span class="co">#&gt;  4 ENSG00000000003 SRR1039513   408</span>
<span class="co">#&gt;  5 ENSG00000000003 SRR1039516  1138</span>
<span class="co">#&gt;  6 ENSG00000000003 SRR1039517  1047</span>
<span class="co">#&gt;  7 ENSG00000000003 SRR1039520   770</span>
<span class="co">#&gt;  8 ENSG00000000003 SRR1039521   572</span>
<span class="co">#&gt;  9 ENSG00000000005 SRR1039508     0</span>
<span class="co">#&gt; 10 ENSG00000000005 SRR1039509     0</span>
<span class="co">#&gt; # … with 512,806 more rows</span></pre></body></html></div>
<p>We can get the gene symbols for these Ensembl gene ids with tidybulk’s ensembl_to_symbol. This works for human and mouse.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">counts_long</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/ensembl_to_symbol-methods.html">ensembl_to_symbol</a></span>(<span class="no">counts_long</span>, <span class="no">GeneID</span>)
<span class="co">#&gt; Joining, by = "GeneID"</span>
<span class="no">counts_long</span>
<span class="co">#&gt; # A tibble: 512,816 x 5</span>
<span class="co">#&gt;    GeneID          sample     count transcript ref_genome</span>
<span class="co">#&gt;    &lt;chr&gt;           &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;      &lt;chr&gt;     </span>
<span class="co">#&gt;  1 ENSG00000000003 SRR1039508   679 TSPAN6     hg38      </span>
<span class="co">#&gt;  2 ENSG00000000003 SRR1039509   448 TSPAN6     hg38      </span>
<span class="co">#&gt;  3 ENSG00000000003 SRR1039512   873 TSPAN6     hg38      </span>
<span class="co">#&gt;  4 ENSG00000000003 SRR1039513   408 TSPAN6     hg38      </span>
<span class="co">#&gt;  5 ENSG00000000003 SRR1039516  1138 TSPAN6     hg38      </span>
<span class="co">#&gt;  6 ENSG00000000003 SRR1039517  1047 TSPAN6     hg38      </span>
<span class="co">#&gt;  7 ENSG00000000003 SRR1039520   770 TSPAN6     hg38      </span>
<span class="co">#&gt;  8 ENSG00000000003 SRR1039521   572 TSPAN6     hg38      </span>
<span class="co">#&gt;  9 ENSG00000000005 SRR1039508     0 TNMD       hg38      </span>
<span class="co">#&gt; 10 ENSG00000000005 SRR1039509     0 TNMD       hg38      </span>
<span class="co">#&gt; # … with 512,806 more rows</span></pre></body></html></div>
<p>We will next extract just the columns we need, sample, transcript, count. To do this we will use the tidyverse pipe <code><a href="https://tibble.tidyverse.org/reference/reexports.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps (as we’ll see later). It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co"># using pipe</span>
<span class="no">counts_long_sel</span> <span class="kw">&lt;-</span> <span class="no">counts_long</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">sample</span>, <span class="no">transcript</span>, <span class="no">count</span>)

<span class="co"># take a look </span>
<span class="no">counts_long_sel</span>
<span class="co">#&gt; # A tibble: 512,816 x 3</span>
<span class="co">#&gt;    sample     transcript count</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;int&gt;</span>
<span class="co">#&gt;  1 SRR1039508 TSPAN6       679</span>
<span class="co">#&gt;  2 SRR1039509 TSPAN6       448</span>
<span class="co">#&gt;  3 SRR1039512 TSPAN6       873</span>
<span class="co">#&gt;  4 SRR1039513 TSPAN6       408</span>
<span class="co">#&gt;  5 SRR1039516 TSPAN6      1138</span>
<span class="co">#&gt;  6 SRR1039517 TSPAN6      1047</span>
<span class="co">#&gt;  7 SRR1039520 TSPAN6       770</span>
<span class="co">#&gt;  8 SRR1039521 TSPAN6       572</span>
<span class="co">#&gt;  9 SRR1039508 TNMD           0</span>
<span class="co">#&gt; 10 SRR1039509 TNMD           0</span>
<span class="co">#&gt; # … with 512,806 more rows</span></pre></body></html></div>
<p>Now we have our counts matrix in the long format we will join it to our sampleinfo so we have information on the samples, what groups they belong to. The join will use all columns with the same name to join. Here we have a column called “sample” in both tables so that will be used to join the two tables.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">counts_annot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(<span class="no">counts_long_sel</span>, <span class="no">sampleinfo</span>)
<span class="co">#&gt; Joining, by = "sample"</span>

<span class="co"># take a look</span>
<span class="no">counts_annot</span>
<span class="co">#&gt; # A tibble: 512,816 x 12</span>
<span class="co">#&gt;    sample transcript count SampleName cell  dex   albut Run   avgLength</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt;</span>
<span class="co">#&gt;  1 SRR10… TSPAN6       679 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  2 SRR10… TSPAN6       448 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  3 SRR10… TSPAN6       873 GSM1275866 N052… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  4 SRR10… TSPAN6       408 GSM1275867 N052… trt   untrt SRR1…        87</span>
<span class="co">#&gt;  5 SRR10… TSPAN6      1138 GSM1275870 N080… untrt untrt SRR1…       120</span>
<span class="co">#&gt;  6 SRR10… TSPAN6      1047 GSM1275871 N080… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  7 SRR10… TSPAN6       770 GSM1275874 N061… untrt untrt SRR1…       101</span>
<span class="co">#&gt;  8 SRR10… TSPAN6       572 GSM1275875 N061… trt   untrt SRR1…        98</span>
<span class="co">#&gt;  9 SRR10… TNMD           0 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; 10 SRR10… TNMD           0 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt; # … with 512,806 more rows, and 3 more variables: Experiment &lt;fct&gt;,</span>
<span class="co">#&gt; #   Sample &lt;fct&gt;, BioSample &lt;fct&gt;</span></pre></body></html></div>
<p>We can shorten the sample names. We can remove the SRR1039 prefix that’s present in all of them, as shorter names can fit better in some of the plots we will create. We can use <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> together with <code><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace()</a></code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">counts_annot_pretty</span> <span class="kw">&lt;-</span> <span class="no">counts_annot</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">sample</span><span class="kw">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span>(<span class="no">sample</span>, <span class="st">"SRR1039"</span>))</pre></body></html></div>
<p>All above operations can be linked, so no temporary variable have to be created.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">counts_annot_pretty</span> <span class="kw">&lt;-</span> <span class="no">counts</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="kw">cols</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">"SRR"</span>), <span class="kw">names_to</span> <span class="kw">=</span> <span class="st">"sample"</span>, <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"count"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/ensembl_to_symbol-methods.html">ensembl_to_symbol</a></span>(<span class="no">GeneID</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">sample</span>, <span class="no">transcript</span>, <span class="no">count</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/join.html">full_join</a></span>(<span class="no">sampleinfo</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">sample</span><span class="kw">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span>(<span class="no">sample</span>, <span class="st">"SRR1039"</span>))
<span class="co">#&gt; Joining, by = "GeneID"</span>
<span class="co">#&gt; Joining, by = "sample"</span></pre></body></html></div>
<p>Now that we have our data in the format we want we will create a tidybulk object, that we can use to perform differential expression analysis with the tidybulk package. For this we need to specify our counts object and the names of the columns that contain our sample ids, our gene identifiers and our counts. Any other columns in the counts object e.g. our Ensembl gene id column will remain at the end.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">#create a 'tt' object</span>
<span class="no">counts_tt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span>(<span class="no">counts_annot_pretty</span>, <span class="no">sample</span>, <span class="no">transcript</span>, <span class="no">count</span>)

<span class="co"># take a look</span>
<span class="no">counts_tt</span>
<span class="co">#&gt; # A tibble: 512,816 x 12</span>
<span class="co">#&gt;    sample transcript count SampleName cell  dex   albut Run   avgLength</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt;</span>
<span class="co">#&gt;  1 508    TSPAN6       679 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  2 509    TSPAN6       448 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  3 512    TSPAN6       873 GSM1275866 N052… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  4 513    TSPAN6       408 GSM1275867 N052… trt   untrt SRR1…        87</span>
<span class="co">#&gt;  5 516    TSPAN6      1138 GSM1275870 N080… untrt untrt SRR1…       120</span>
<span class="co">#&gt;  6 517    TSPAN6      1047 GSM1275871 N080… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  7 520    TSPAN6       770 GSM1275874 N061… untrt untrt SRR1…       101</span>
<span class="co">#&gt;  8 521    TSPAN6       572 GSM1275875 N061… trt   untrt SRR1…        98</span>
<span class="co">#&gt;  9 508    TNMD           0 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; 10 509    TNMD           0 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt; # … with 512,806 more rows, and 3 more variables: Experiment &lt;fct&gt;,</span>
<span class="co">#&gt; #   Sample &lt;fct&gt;, BioSample &lt;fct&gt;</span></pre></body></html></div>
<p>Some gene symbols are not unique, they map to more than one gene id. We need to remove this redundancy and we can do that with tidybulk function <code><a href="https://rdrr.io/pkg/tidybulk/man/aggregate_duplicates-methods.html">aggregate_duplicates()</a></code>. By default it will aggregate duplicate gene symbols summing their counts.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># get rid of duplicated gene symbols and drop any NAs</span>
<span class="no">counts_aggr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/aggregate_duplicates-methods.html">aggregate_duplicates</a></span>(<span class="no">counts_tt</span>)
<span class="co">#&gt; Warning in aggregate_duplicated_transcripts_bulk(.data, .sample = !!.sample, :</span>
<span class="co">#&gt; tidybulk says: for aggregation fctors and logical columns were converted to</span>
<span class="co">#&gt; character</span>
<span class="co">#&gt; Converted to characters</span>
<span class="co">#&gt; factorfactorfactorfactorfactorfactorfactorfactor</span>
<span class="no">counts_aggr</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html">drop_na</a></span>(<span class="no">counts_aggr</span>, <span class="no">transcript</span>)</pre></body></html></div>
<p>We can check how many counts we have for each sample by making a bar plot. This helps us see whether there are any major discrepancies between the samples more easily. We use weight= to sum up the counts for each sample.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co"># make barplot of counts</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">counts_aggr</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">weight</span><span class="kw">=</span><span class="no">count</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">dex</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-14-1.png" width="700"> We can also easily view by cell line (or any other variable that’s a column in our dataset) simply by changing fill=.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># make barplot of counts</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">counts_aggr</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">weight</span><span class="kw">=</span><span class="no">count</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">cell</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>The bar plots show us there are 15-30 million counts per sample.</p>
</div>
</div>
<div id="filtering-lowly-expressed-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering-lowly-expressed-genes" class="anchor"></a>Filtering lowly expressed genes</h1>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>tidybulk can automatically filter out lowly expressed genes. It uses the edgeR filterByExpr function described <a href="https://f1000research.com/articles/5-1408">here</a>. This will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is 2 samples. tidybulk performs this filtering in the functions we will use <code><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance()</a></code> and <code><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance()</a></code> and we can take a look at it with <code><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant()</a></code>.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># Filtering out lowly abundant genes </span>
<span class="no">counts_filt</span> <span class="kw">&lt;-</span> <span class="no">counts_aggr</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_abundant-methods.html">keep_abundant</a></span>(<span class="kw">factor_of_interest</span> <span class="kw">=</span> <span class="no">dex</span>)

<span class="co"># take a look</span>
<span class="no">counts_filt</span>
<span class="co">#&gt; # A tibble: 116,600 x 13</span>
<span class="co">#&gt;    sample transcript count SampleName cell  dex   albut Run   avgLength</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 508    TSPAN6       679 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  2 509    TSPAN6       448 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  3 512    TSPAN6       873 GSM1275866 N052… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  4 513    TSPAN6       408 GSM1275867 N052… trt   untrt SRR1…        87</span>
<span class="co">#&gt;  5 516    TSPAN6      1138 GSM1275870 N080… untrt untrt SRR1…       120</span>
<span class="co">#&gt;  6 517    TSPAN6      1047 GSM1275871 N080… trt   untrt SRR1…       126</span>
<span class="co">#&gt;  7 520    TSPAN6       770 GSM1275874 N061… untrt untrt SRR1…       101</span>
<span class="co">#&gt;  8 521    TSPAN6       572 GSM1275875 N061… trt   untrt SRR1…        98</span>
<span class="co">#&gt;  9 508    DPM1         467 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; 10 509    DPM1         515 GSM1275863 N613… trt   untrt SRR1…       126</span>
<span class="co">#&gt; # … with 116,590 more rows, and 4 more variables: Experiment &lt;chr&gt;,</span>
<span class="co">#&gt; #   Sample &lt;chr&gt;, BioSample &lt;chr&gt;, `merged transcripts` &lt;dbl&gt;</span></pre></body></html></div>
<p>We can create density plots to view the distributions of the counts for the samples. This is also a quality check to see if the samples look similar and that none look majorly different. Note we need to log the counts which we can do by using <code><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10()</a></code> We need to add a small offset (1) to the counts to avoid taking log of zero.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># density plot after filtering </span>
<span class="no">counts_filt</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">count</span> + <span class="fl">1</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">color</span><span class="kw">=</span><span class="no">dex</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-17-1.png" width="700"> These samples all look pretty similar, none are majorly different.</p>
<p>We can count how many genes there are after filtering.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">counts_filt</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">transcript</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span>()
<span class="co">#&gt; [1] 14575</span></pre></body></html></div>
<div id="exercise" class="section level4">
<h4 class="hasAnchor">
<a href="#exercise" class="anchor"></a>Exercise</h4>
<p>Adapt the code above to create a density plot of the counts before filtering lowly expressed genes. How does it compare it to the density plot above? Count how many genes there were before filtering.</p>
</div>
</div>
<div id="normalisation-for-sequencing-depth-and-composition" class="section level1">
<h1 class="hasAnchor">
<a href="#normalisation-for-sequencing-depth-and-composition" class="anchor"></a>Normalisation for sequencing depth and composition</h1>
<p>TMM normalisation is performed to eliminate composition biases between libraries <span class="citation">[@robinson2010tmm]</span>. This generates a set of normalisation factors, where the product of these factors and the library sizes defines the effective library size. TMM normalisation (and most scaling normalisation methods) scale relative to one sample. In the tidybulk package the function <code><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance()</a></code> generates scaled counts, it will also perform the filtering.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co"># Scaling counts for library size and composition bias</span>
<span class="no">counts_scaled</span> <span class="kw">&lt;-</span> <span class="no">counts_filt</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span>(<span class="kw">factor_of_interest</span> <span class="kw">=</span> <span class="no">dex</span>)

<span class="co"># take a look</span>
<span class="no">counts_scaled</span>
<span class="co">#&gt; # A tibble: 116,600 x 17</span>
<span class="co">#&gt;    sample transcript count SampleName cell  dex   albut Run   avgLength</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 508    A1BG-AS1      38 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  2 508    A2M        20085 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  3 508    A2M-AS1       49 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  4 508    A4GALT       834 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  5 508    AAAS         780 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  6 508    AACS         399 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  7 508    AADAT        171 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  8 508    AAGAB        680 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  9 508    AAK1         287 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; 10 508    AAMDC         89 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; # … with 116,590 more rows, and 8 more variables: Experiment &lt;chr&gt;,</span>
<span class="co">#&gt; #   Sample &lt;chr&gt;, BioSample &lt;chr&gt;, `merged transcripts` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;</span></pre></body></html></div>
<p>After we run <code><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance()</a></code> we should see some columns have been added at the end. We have a column called <code>lowly_abundant</code> that indicates whether the gene has been filtered due to being lowly expressed. FALSE means the gene wasn’t filtered, TRUE means it was. The <code>count_scaled</code> column contains the scaled counts.</p>
<p>We can also create box plots to check the distributions of the counts in the samples. We can add a line through the median with to help us see how similar (or not) the distributions are.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co"># box plot after scaling</span>
<span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">lowly_abundant</span> <span class="kw">==</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">count_scaled</span> + <span class="fl">1</span>, <span class="kw">fill</span><span class="kw">=</span><span class="no">dex</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">yintercept</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span>(<span class="no">count_scaled</span> + <span class="fl">1</span>)), <span class="kw">colour</span> <span class="kw">=</span> <span class="st">'red'</span>, <span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-20-1.png" width="700"> These samples all look pretty similar, none are majorly different.</p>
<div id="exercise-1" class="section level4">
<h4 class="hasAnchor">
<a href="#exercise-1" class="anchor"></a>Exercise</h4>
<p>Adapt the code above to create box plots of the counts before scaling. How do they compare to the box plots above?</p>
</div>
</div>
<div id="quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h1>
<div id="multidimensional-scaling-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#multidimensional-scaling-plots" class="anchor"></a>Multidimensional scaling plots</h2>
<p>By far, one of the most important plots we make when we analyse RNA-Seq data are MDS plots. An MDS plot is a visualisation of a principal components analysis, which determines the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions()</a></code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co"># get MDS dimensions</span>
<span class="no">counts_scal_MDS</span> <span class="kw">&lt;-</span>
  <span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span>(<span class="kw">method</span><span class="kw">=</span><span class="st">"MDS"</span>, <span class="kw">.dims</span> <span class="kw">=</span> <span class="fl">2</span>)
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "tt_internals")$MDS`</span>

<span class="co"># take a look</span>
<span class="no">counts_scal_MDS</span>
<span class="co">#&gt; # A tibble: 116,600 x 19</span>
<span class="co">#&gt;    sample transcript count SampleName cell  dex   albut Run   avgLength</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 508    A1BG-AS1      38 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  2 508    A2M        20085 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  3 508    A2M-AS1       49 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  4 508    A4GALT       834 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  5 508    AAAS         780 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  6 508    AACS         399 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  7 508    AADAT        171 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  8 508    AAGAB        680 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt;  9 508    AAK1         287 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; 10 508    AAMDC         89 GSM1275862 N613… untrt untrt SRR1…       126</span>
<span class="co">#&gt; # … with 116,590 more rows, and 10 more variables: Experiment &lt;chr&gt;,</span>
<span class="co">#&gt; #   Sample &lt;chr&gt;, BioSample &lt;chr&gt;, `merged transcripts` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   count_scaled &lt;dbl&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;,</span>
<span class="co">#&gt; #   Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;</span></pre></body></html></div>
<p>Then we can select just the dimensions for the samples.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="co"># get the dimensions with all metadata</span>
<span class="no">MDSdims</span> <span class="kw">&lt;-</span>
    <span class="no">counts_scal_MDS</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span>()

<span class="co"># take a look</span>
<span class="no">MDSdims</span>
<span class="co">#&gt; # A tibble: 8 x 15</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt; </span>
<span class="co">#&gt; 1 508    GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 509    GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 512    GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 513    GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 516    GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 517    GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 520    GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 521    GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 6 more variables: BioSample &lt;chr&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;,</span>
<span class="co">#&gt; #   lowly_abundant &lt;lgl&gt;, Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;</span></pre></body></html></div>
<p>Next we can plot the MDS dimensions as a scatterplot. We’ll colour by the conditions to see if the replicates group together.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="co"># MDS plot</span>
<span class="no">MDSdims</span> <span class="kw">%&gt;%</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">Dim1</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">Dim2</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">dex</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div id="exercise-2" class="section level4">
<h4 class="hasAnchor">
<a href="#exercise-2" class="anchor"></a>Exercise</h4>
<p>Colour the MDS plot i) by sample, ii) by cell line. Try using <code>shape=</code> inside the aes() with the dex and/or cell variables. Discuss what is the greatest source of variation in the data (i.e. what does dimension 1 represent)? What is the second greatest source of variation in the data?</p>
</div>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h2>
<p>An alternative to MDS for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use to extract the 500 most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">counts_scal_MDS</span> <span class="kw">%&gt;%</span>

    <span class="co"># extract 500 most variable genes</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span>(<span class="kw">.sample</span> <span class="kw">=</span> <span class="no">sample</span>, <span class="kw">.transcript</span> <span class="kw">=</span> <span class="no">transcript</span>, <span class="kw">.abundance</span> <span class="kw">=</span> <span class="no">count_scaled</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">500</span>) <span class="kw">%&gt;%</span>

    <span class="co"># create heatmap</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span>(
          <span class="kw">.column</span> <span class="kw">=</span> <span class="no">sample</span>,
          <span class="kw">.row</span> <span class="kw">=</span> <span class="no">transcript</span>,
          <span class="kw">.value</span> <span class="kw">=</span> <span class="no">count_scaled</span>,
          <span class="kw">annotation</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">dex</span>, <span class="no">cell</span>),
          <span class="kw">transform</span> <span class="kw">=</span> <span class="no">log1p</span>
      )
<span class="co">#&gt; Getting the 500 most variable genes</span></pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-24-1.png" width="768"></p>
</div>
</div>
<div id="differential-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h1>
<p>Now that we are happy that the data looks good, we can continue to testing for differentially expressed genes. We will use the <code><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance()</a></code> from tidybulk which uses edgeR to perform the differential expression analysis. We give <code><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance()</a></code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line we could use the formula <code>0 + dex</code>, however, each treated and untreated sample is from a different celline so we add the cell line as an additional factor <code>0 + dex + cell</code>. We also provide the names of the groups we want to compare/contrast to .contrasts (e.g. .contrasts = c(“dextreat - dexuntreat”)). <code><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance()</a></code> will perform the filtering of lowly expressed genes as described before. The results will be joined to our counts for every sample. If we just want a table of differentially expressed genes we can use pivot_transcript.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">counts_de</span> <span class="kw">&lt;-</span> <span class="no">counts_scal_MDS</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span>(
      <span class="kw">.formula</span> <span class="kw">=</span> ~ <span class="fl">0</span> + <span class="no">dex</span> + <span class="no">cell</span>,
      <span class="kw">.contrasts</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dextrt - dexuntrt"</span>))
<span class="co">#&gt; Warning in get_differential_transcript_abundance_bulk(.data, .formula, .sample</span>
<span class="co">#&gt; = !!.sample, : tidybulk says: You have less than two replicated for each</span>
<span class="co">#&gt; factorial condition</span>
<span class="co">#&gt; tidybulk says: The design column names are "dextrt, dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; Joining, by = "transcript"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (glmFit) do `attr(..., "tt_internals")$edgeR`</span>
<span class="co">#&gt; Joining, by = c("transcript", "lowly_abundant")</span>

<span class="co"># take a look</span>
<span class="no">counts_de</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>()
<span class="co">#&gt; # A tibble: 14,575 x 10</span>
<span class="co">#&gt;    transcript albut `merged transcr… lowly_abundant `logFC_dextrt -…</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;lgl&gt;                     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 A1BG-AS1   untrt                1 FALSE                    0.503 </span>
<span class="co">#&gt;  2 A2M        untrt                1 FALSE                    0.523 </span>
<span class="co">#&gt;  3 A2M-AS1    untrt                1 FALSE                   -0.344 </span>
<span class="co">#&gt;  4 A4GALT     untrt                1 FALSE                    0.511 </span>
<span class="co">#&gt;  5 AAAS       untrt                1 FALSE                   -0.0338</span>
<span class="co">#&gt;  6 AACS       untrt                1 FALSE                   -0.199 </span>
<span class="co">#&gt;  7 AADAT      untrt                1 FALSE                   -0.648 </span>
<span class="co">#&gt;  8 AAGAB      untrt                1 FALSE                   -0.173 </span>
<span class="co">#&gt;  9 AAK1       untrt                1 FALSE                   -0.200 </span>
<span class="co">#&gt; 10 AAMDC      untrt                1 FALSE                    0.438 </span>
<span class="co">#&gt; # … with 14,565 more rows, and 5 more variables: `logCPM_dextrt -</span>
<span class="co">#&gt; #   dexuntrt` &lt;dbl&gt;, `LR_dextrt - dexuntrt` &lt;dbl&gt;, `PValue_dextrt -</span>
<span class="co">#&gt; #   dexuntrt` &lt;dbl&gt;, `FDR_dextrt - dexuntrt` &lt;dbl&gt;, `significant_dextrt -</span>
<span class="co">#&gt; #   dexuntrt` &lt;lgl&gt;</span></pre></body></html></div>
<p>Now we have columns with our logFC and FDR P values. There is also a column called significant that indicates if the gene has FDR &lt; 0.05.</p>
<div id="table-of-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#table-of-differentially-expressed-genes" class="anchor"></a>Table of differentially expressed genes</h3>
<p>We can write out our differentialy expressed genes to a file that can be loaded into e.g. Excel. <code><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv()</a></code> will create a tab-separated file.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co"># save results</span>
<span class="no">counts_de</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>( <span class="st">"de_results.tsv"</span>)</pre></body></html></div>
<p>We only have one contrast here so we can remove the suffix <code>_conditionbpreg - conditionblact</code> from the column headers, to make the names shorter to use in the rest of the commands here.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">&lt;-</span>
    <span class="no">counts_de</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_at</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/vars.html">vars</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">contains</a></span>(<span class="st">"dex"</span>)),
            ~<span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span>(<span class="no">.</span>, <span class="st">"_dextrt - dexuntrt"</span>, <span class="st">""</span>))</pre></body></html></div>
</div>
<div id="counting-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-differentially-expressed-genes" class="anchor"></a>Counting differentially expressed genes</h3>
<p>We can count how many differentially expressed genes there are using the significant column.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">significant</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">transcript</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span>()
<span class="co">#&gt; [1] 5247</span></pre></body></html></div>
<div id="exercise-3" class="section level4">
<h4 class="hasAnchor">
<a href="#exercise-3" class="anchor"></a>Exercise</h4>
<p>Count how many upregulated DE genes there are.<br>
Count how many downregulated DE genes there are.</p>
</div>
</div>
<div id="extracting-top-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#extracting-top-differentially-expressed-genes" class="anchor"></a>Extracting top differentially expressed genes</h3>
<p>We can get the top genes by P value with</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">significant</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">PValue</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">6</span>)
<span class="co">#&gt; # A tibble: 6 x 10</span>
<span class="co">#&gt;   transcript albut `merged transcr… lowly_abundant logFC logCPM    LR    PValue</span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;            &lt;dbl&gt; &lt;lgl&gt;          &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ZBTB16     untrt                1 FALSE           7.14   4.16 1214. 6.75e-266</span>
<span class="co">#&gt; 2 CACNB2     untrt                1 FALSE           3.27   4.52  746. 2.52e-164</span>
<span class="co">#&gt; 3 SPARCL1    untrt                1 FALSE           4.55   5.54  635. 4.66e-140</span>
<span class="co">#&gt; 4 ANGPTL7    untrt                1 FALSE           5.66   3.51  630. 6.23e-139</span>
<span class="co">#&gt; 5 FAM107A    untrt                1 FALSE           4.71   2.79  560. 6.61e-124</span>
<span class="co">#&gt; 6 STEAP4     untrt                1 FALSE           5.18   3.66  547. 5.36e-121</span>
<span class="co">#&gt; # … with 2 more variables: FDR &lt;dbl&gt;, significant &lt;lgl&gt;</span></pre></body></html></div>
<p>We can extract the symbols for these top genes with</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="no">topgenes</span> <span class="kw">&lt;-</span>
    <span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">significant</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">PValue</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">6</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span>(<span class="no">transcript</span>)</pre></body></html></div>
</div>
</div>
<div id="plots-after-testing-for-de" class="section level1">
<h1 class="hasAnchor">
<a href="#plots-after-testing-for-de" class="anchor"></a>Plots after testing for DE</h1>
<p>Let’s make a few plots to make sure everything looks good and that we haven’t made a mistake in the analysis. Genome-wide plots that are useful for checking are MA plots and volcano plots. We can also use stripcharts and heatmaps to visualise groups of genes.</p>
<div id="ma-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#ma-plots" class="anchor"></a>MA plots</h3>
<p>MA plots enable us to visualise <strong>amount</strong> of expression (logCPM) versus logFC. Highly expressed genes are towards the right of the plot. We can also colour significant genes (e.g. genes with FDR &lt; 0.05)</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co"># maplot, minimal</span>
<span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">lowly_abundant</span> <span class="kw">==</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">logCPM</span>, <span class="kw">y</span><span class="kw">=</span>-<span class="no">logFC</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">significant</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>A more informative plot, integrating some of the packages in tidyverse</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>

    <span class="co"># Subset data</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">significant</span> <span class="kw">=</span> <span class="no">FDR</span><span class="kw">&lt;</span><span class="fl">0.05</span> <span class="kw">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;=</span><span class="fl">2</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">transcript</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;=</span><span class="fl">7</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">transcript</span>), <span class="fl">NA</span>)) <span class="kw">%&gt;%</span>

    <span class="co"># Plot</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">logCPM</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">logFC</span>, <span class="kw">label</span><span class="kw">=</span><span class="no">transcript</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">size</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="no">significant</span>)) +
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"#e11f28"</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span>(<span class="kw">range</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">2</span>)) +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()
<span class="co">#&gt; Warning: Using size for a discrete variable is not advised.</span>
<span class="co">#&gt; Warning: Using alpha for a discrete variable is not advised.</span>
<span class="co">#&gt; Warning: Removed 14573 rows containing missing values (geom_text_repel).</span></pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
<div id="volcano-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h3>
<p>Volcano plots enable us to visualise <strong>significance</strong> of expression (logCPM) versus logFC. Highly significant genes are towards the top of the plot. We can also colour significant genes (e.g. genes with FDR &lt; 0.05)</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="co"># volcanoplot, minimal</span>
<span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">lowly_abundant</span> <span class="kw">==</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">logFC</span>, <span class="kw">y</span><span class="kw">=</span>-<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">PValue</span>), <span class="kw">colour</span><span class="kw">=</span><span class="no">significant</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-33-1.png" width="700"> To see how to make more complicated volcano plots, including how to label genes in the plot, see the volcano plot tutorial <a href="https://pmacdasci.github.io/r-intro-tidyverse/volcanoplot.html">here</a>. More complicated MA plots could also be made in a similar way.</p>
</div>
<div id="stripcharts" class="section level3">
<h3 class="hasAnchor">
<a href="#stripcharts" class="anchor"></a>Stripcharts</h3>
<p>In addition to the genome-wide plots already discussed, it is recommended to have a look at the expression levels of the individual samples for the genes of interest, before following up on the DE genes with further lab work. We can use stripcharts and heatmaps to do this. These will help show if expression is consistent amongst replicates in the groups.</p>
<p>With stripcharts we can see if replicates tend to group together and how the expression compares to the other groups.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>

    <span class="co"># extract counts for top DE genes</span>
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">transcript</span> <span class="kw">%in%</span> <span class="no">topgenes</span>) <span class="kw">%&gt;%</span>

    <span class="co"># make stripchart</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">dex</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">count_scaled</span> + <span class="fl">1</span>, <span class="kw">colour</span> <span class="kw">=</span> <span class="no">dex</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">transcript</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
</div>
<div id="heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#heatmaps" class="anchor"></a>Heatmaps</h3>
<p>We can create heatmaps for the most differentially expressed genes. For example we could select the genes with FDR &lt; 0.05 and a logFC change of 4 and make a heatmap of those.</p>
<p>Then we can create the heatmap selecting genes by applying filters for e.g. FDR and logFC.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">FDR</span> <span class="kw">&lt;</span> <span class="fl">0.05</span> <span class="kw">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;</span> <span class="fl">4</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span>(
        <span class="kw">.column</span> <span class="kw">=</span> <span class="no">sample</span>,
        <span class="kw">.row</span> <span class="kw">=</span> <span class="no">transcript</span>,
        <span class="kw">.value</span> <span class="kw">=</span> <span class="no">count_scaled</span>,
        <span class="kw">annotation</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">dex</span>, <span class="no">cell</span>),
        <span class="kw">transform</span> <span class="kw">=</span> <span class="no">log1p</span>
    )</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-35-1.png" width="768"></p>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo()</a></code></p>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.0 (2020-04-24)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt;  [1] parallel  stats4    grid      stats     graphics  grDevices utils    </span>
<span class="co">#&gt;  [8] datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] airway_1.9.0                SummarizedExperiment_1.19.4</span>
<span class="co">#&gt;  [3] DelayedArray_0.15.1         matrixStats_0.56.0         </span>
<span class="co">#&gt;  [5] Biobase_2.49.0              GenomicRanges_1.41.1       </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.25.0         IRanges_2.23.6             </span>
<span class="co">#&gt;  [9] S4Vectors_0.27.9            BiocGenerics_0.35.2        </span>
<span class="co">#&gt; [11] tidybulk_1.1.0              tidyHeatmap_0.99.18        </span>
<span class="co">#&gt; [13] ComplexHeatmap_2.5.3        ggrepel_0.8.2              </span>
<span class="co">#&gt; [15] ggplot2_3.3.0               stringr_1.4.0              </span>
<span class="co">#&gt; [17] readr_1.3.1                 tidyr_1.1.0                </span>
<span class="co">#&gt; [19] dplyr_0.8.5                 tibble_3.0.1               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] viridis_0.5.1          edgeR_3.31.0           viridisLite_0.3.0     </span>
<span class="co">#&gt;  [4] assertthat_0.2.1       GenomeInfoDbData_1.2.3 yaml_2.2.1            </span>
<span class="co">#&gt;  [7] pillar_1.4.4           backports_1.1.7        lattice_0.20-41       </span>
<span class="co">#&gt; [10] limma_3.45.0           glue_1.4.1             digest_0.6.25         </span>
<span class="co">#&gt; [13] RColorBrewer_1.1-2     XVector_0.29.1         colorspace_1.4-1      </span>
<span class="co">#&gt; [16] htmltools_0.4.0        preprocessCore_1.51.0  Matrix_1.2-18         </span>
<span class="co">#&gt; [19] pkgconfig_2.0.3        GetoptLong_0.1.8       zlibbioc_1.35.0       </span>
<span class="co">#&gt; [22] purrr_0.3.4            scales_1.1.1           farver_2.0.3          </span>
<span class="co">#&gt; [25] ellipsis_0.3.1         withr_2.2.0            cli_2.0.2             </span>
<span class="co">#&gt; [28] magrittr_1.5           crayon_1.3.4           memoise_1.1.0         </span>
<span class="co">#&gt; [31] evaluate_0.14          fs_1.4.1               fansi_0.4.1           </span>
<span class="co">#&gt; [34] MASS_7.3-51.5          tools_4.0.0            hms_0.5.3             </span>
<span class="co">#&gt; [37] GlobalOptions_0.1.1    lifecycle_0.2.0        locfit_1.5-9.4        </span>
<span class="co">#&gt; [40] munsell_0.5.0          cluster_2.1.0          compiler_4.0.0        </span>
<span class="co">#&gt; [43] pkgdown_1.5.1          rlang_0.4.6            RCurl_1.98-1.2        </span>
<span class="co">#&gt; [46] rjson_0.2.20           circlize_0.4.9         labeling_0.3          </span>
<span class="co">#&gt; [49] bitops_1.0-6           rmarkdown_2.1          gtable_0.3.0          </span>
<span class="co">#&gt; [52] R6_2.4.1               gridExtra_2.3          knitr_1.28            </span>
<span class="co">#&gt; [55] utf8_1.1.4             clue_0.3-57            rprojroot_1.3-2       </span>
<span class="co">#&gt; [58] shape_1.4.4            desc_1.2.0             stringi_1.4.6         </span>
<span class="co">#&gt; [61] Rcpp_1.0.4.6           vctrs_0.3.0            png_0.1-7             </span>
<span class="co">#&gt; [64] tidyselect_1.1.0       xfun_0.14</span></pre></body></html></div>
</div>
<div id="key-points" class="section level1">
<h1 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h1>
<ul>
<li>RNA-Seq data can be analysed in a ‘tidy’ way using the packages tidyverse, tidybulk and tidyHeatmap</li>
<li>Key steps in an RNA-Seq analysis are filtering lowly expressed genes, normalisation for sequencing depth and composition, and testing for differential expression</li>
<li>MDS plots are very important for examining the quality of the data</li>
<li>Other useful plots for assessing RNA-Seq data are bar plots, density plots, box plots, MA plots, volcano plots, stripcharts and heatmaps. These can all be made in a ‘tidy’ way.</li>
</ul>
</div>
<div id="further-reading" class="section level1">
<h1 class="hasAnchor">
<a href="#further-reading" class="anchor"></a>Further Reading</h1>
<p><a href="https://f1000research.com/articles/5-1408">RNA-Seq analysis is easy as 1-2-3 with limma, Glimma and edgeR</a><br><a href="http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html">RNA-Seq analysis in R</a></p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:maria.doyle@petermac.org" class="email">maria.doyle@petermac.org</a><a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p><a href="mailto:mangiola.s@wehi.edu.au" class="email">mangiola.s@wehi.edu.au</a><a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
