<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Tidy Transcriptomics introduction to RNA-Seq analyses • tidytranscriptomics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A Tidy Transcriptomics introduction to RNA-Seq analyses">
<meta property="og:description" content="tidytranscriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidytranscriptomics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tidytranscriptomics.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/bioc_2020_tidytranscriptomics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidytranscriptomics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="tidytranscriptomics_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="tidytranscriptomics_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="tidytranscriptomics_files/typedarray-0.1/typedarray.min.js"></script><script src="tidytranscriptomics_files/jquery-1.11.3/jquery.min.js"></script><link href="tidytranscriptomics_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="tidytranscriptomics_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="tidytranscriptomics_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="tidytranscriptomics_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A Tidy Transcriptomics introduction to RNA-Seq analyses</h1>
                        <h4 class="author">Maria Doyle<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Stefano Mangiola<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/bioc_2020_tidytranscriptomics/blob/master/vignettes/tidytranscriptomics.Rmd"><code>vignettes/tidytranscriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>tidytranscriptomics.Rmd</code></div>

    </div>

    
    
<div id="workshop-description" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h2>
<p>This workshop will present how to perform analysis of RNA sequencing data following the tidy data paradigm. The tidy data paradigm provides a standard way to organise data values within a dataset, where each variable is a column, each observation is a row, and data is manipulated using an easy-to-understand vocabulary. Most importantly, the data structure remains consistent across manipulation and analysis functions.</p>
<p>This can be achieved for RNA sequencing data with the <a href="github.com/stemangiola/tidybulk">tidybulk</a>, <a href="github.com/stemangiola/tidyHeatmap">tidyHeatmap</a> and tidyverse packages. The package <a href="github.com/stemangiola/tidybulk">tidybulk</a> provides a tidy data structure and a modular framework for bulk transcriptional analyses. tidyHeatmap provides a tidy implementation of ComplexHeatmap. These packages are part of the tidytranscriptomics suite that introduces a tidy approach to RNA sequencing data.</p>
<p>The topics presented in this workshop will be</p>
<ul>
<li>Data exploration</li>
<li>Data dimensionality reduction and clustering</li>
<li>Differential gene expression analysis</li>
<li>Data visualisation</li>
</ul>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of RStudio</li>
<li>Familiarity with tidyverse syntax</li>
</ul>
<p>Recommended Background Reading <a href="https://melbournebioinformatics.github.io/r-intro-biologists/intro_r_biologists.html">Introduction to R for Biologists</a></p>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Students will be expected to participate in the workshop in a hands-on way, following along with the code provided and performing exercises.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li>tidyverse</li>
<li>tidybulk</li>
<li>tidyHeatmap</li>
<li>edgeR</li>
<li>ggrepel</li>
<li>airway</li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data preprocessing</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Data dimensionality reduction and clustering</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Differential gene expression</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Data visualisation</td>
<td>20m</td>
</tr>
</tbody>
</table>
</div>
<div id="workshop-goals-and-objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-goals-and-objectives" class="anchor"></a>Workshop goals and objectives</h1>
<p>In exploring and analysing RNA sequencing data, there are a number of key concepts, such as filtering, scaling, dimensionality reduction, hypothesis testing, clustering and visualisation, that need to be understood. These concepts can be intuitively explained to new users, however, (i) the use of a heterogeneous vocabulary and jargon by methodologies/algorithms/packages, (ii) the complexity of data wrangling, and (iii) the coding burden, impede effective learning of the statistics and biology underlying an informed RNA sequencing analysis.</p>
<p>The tidytranscriptomics approach to RNA sequencing data analysis abstracts out the coding-related complexity and provides tools that use an intuitive and jargon-free vocabulary, enabling focus on the statistical and biological challenges.</p>
<div id="learning-goals" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning goals</h2>
<ul>
<li>To understand the key concepts and steps of bulk RNA sequencing data analysis</li>
<li>To approach data representation and analysis though a tidy data paradigm, integrating tidyverse with tidybulk and tidyHeatmap.</li>
</ul>
</div>
<div id="learning-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning objectives</h2>
<ul>
<li>Recall the key concepts of RNA sequencing data analysis</li>
<li>Apply the concepts to publicly available data</li>
<li>Create plots that summarise the information content of the data and analysis results</li>
</ul>
</div>
<div id="acknowledgements" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h2>
<p>This material is adapted from an R for RNA-Seq workshop first run <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>.</p>
<p><img src="../../../../tmp/RtmpkuU2Az/temp_libpath361398a16a7/tidytranscriptomics/vignettes/tidybulk_logo.png" width="501"></p>
</div>
</div>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<ul>
<li>Setting up the data</li>
<li>Filtering lowly expressed genes</li>
<li>Scaling of counts</li>
<li>Quality control plots</li>
<li>Differential expression analysis</li>
<li>Visualisation (MA plot, Volcano plot)</li>
</ul>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA-Seq to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA-Seq experiment. Analysing an RNAseq experiment begins with sequencing reads. These are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and getting stuck into analysis.</p>
<p>First, let’s load all the packages we will need to analyse the data.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># load libraries</span>

<span class="co"># tidyverse core packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tibble</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">readr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">stringr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)

<span class="co"># tidyverse-friendly packages</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyHeatmap</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidybulk</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggrepel</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">plotly</span>)

<span class="co"># dataset</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">airway</span>)</pre></body></html></div>
<div id="airway-rna-seq-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#airway-rna-seq-dataset" class="anchor"></a>Airway RNA-seq dataset</h2>
<p>Here we will perform RNA-Seq analysis using the data from the <em>airway</em> package. It has 8 samples, a sample treated with dex and an untreated control sample for 4 cell lines.</p>
</div>
<div id="setting-up-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up-the-data" class="anchor"></a>Setting up the data</h2>
<p>The airway RNA-Seq data is stored as a <em>RangedSummarizedExperiment</em> object. We’ll convert the SummarizedExperiment object into a tidybulk tibble, which is the tidyverse table format.</p>
<p>In this workshop we will be using the tidyverse pipe <code><a href="https://tibble.tidyverse.org/reference/reexports.html">%&gt;%</a></code>. This ‘pipes’ the output from the command on the left into the command on the right/below. Using the pipe is not essential but it reduces the amount of code we need to write when we have multiple steps (as we’ll see later). It also can make the steps clearer and easier to see. For more details on the pipe see <a href="https://r4ds.had.co.nz/pipes.html">here</a>.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># load airway RNA-Seq data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">airway</span>)

<span class="co"># convert to tidybulk format</span>
<span class="no">counts_airway</span> <span class="kw">&lt;-</span>
    <span class="no">airway</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span>()</pre></body></html></div>
<p>You can type the name of the object to view the first few lines and to see how many rows and columns it has.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">counts_airway</span>
<span class="co">#&gt; # A tibble: 512,816 x 12</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;fct&gt;   &lt;fct&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… SRR10…    679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… SRR10…    467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… SRR10…    260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… SRR10…     60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… SRR10…      0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… SRR10…   3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… SRR10…   1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… SRR10…    519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… SRR10…    394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 2 more variables: Sample &lt;fct&gt;, BioSample &lt;fct&gt;</span></pre></body></html></div>
<p>The <code>counts_tt</code> object contains information about genes and samples, the first column has the Ensembl gene identifier, the second column has the sample identifier and the third column has the gene transcription abundance expressed in number of reads aligning to the gene in each experimental sample. The remaining columns include sample-wise information. The dex column tells us whether the samples are treated or untreated and the cell column tells us what cell line they are from.</p>
<p>We can shorten the sample names. We can remove the SRR1039 prefix that’s present in all of them, as shorter names can fit better in some of the plots we will create. We can use <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code> together with <code><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace()</a></code> to remove the SRR1039 string from the sample column.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">counts_format</span> <span class="kw">&lt;-</span> <span class="no">counts_airway</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">sample</span><span class="kw">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span>(<span class="no">sample</span>, <span class="st">"SRR1039"</span>))</pre></body></html></div>
</div>
<div id="adding-gene-symbols" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-gene-symbols" class="anchor"></a>Adding gene symbols</h2>
<p>We can get the gene symbols for these Ensembl gene ids with tidybulk’s <code>ensembl_to_symbol</code>. This works for human and mouse.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">counts_tt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/ensembl_to_symbol-methods.html">ensembl_to_symbol</a></span>(<span class="no">counts_format</span>, <span class="no">feature</span>)
<span class="no">counts_tt</span>
<span class="co">#&gt; # A tibble: 512,816 x 14</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 4 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, transcript &lt;chr&gt;, ref_genome &lt;chr&gt;</span></pre></body></html></div>
<p>With tidyverse, all above operations can be linked with the <code><a href="https://tibble.tidyverse.org/reference/reexports.html">%&gt;%</a></code>, as shown below. This has the benefits that * no temporary variables need to be created * less typing is required * the steps can be seen more clearly.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">counts_tt</span> <span class="kw">&lt;-</span>
    <span class="no">airway</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">sample</span><span class="kw">=</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span>(<span class="no">sample</span>, <span class="st">"SRR1039"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/ensembl_to_symbol-methods.html">ensembl_to_symbol</a></span>(<span class="no">feature</span>)</pre></body></html></div>
<p>From this tidybulk tibble, we can perform differential expression analysis with the tidybulk package.</p>
</div>
</div>
<div id="filtering-lowly-expressed-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#filtering-lowly-expressed-genes" class="anchor"></a>Filtering lowly expressed genes</h1>
<p>Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>
<p>With tidybulk, it is not really necessary to explicitly filter lowly transcribed genes, as all calculations (e.g., scaling, removal of unwanted variation, differential expression testing) are performed on abundantly transcribed genes; although in case of scaling, the scaling is applied back to all genes in the dataset. Tidybulk can use the <em>edgeR</em> <code>filterByExpr</code> function described <a href="https://f1000research.com/articles/5-1408">here</a>. By default, this will keep genes with ~10 counts in a minimum number of samples, the number of the samples in the smallest group. In this dataset the smallest group size is 2 samples. tidybulk performs this filtering in the functions we will use <code>scale_abundance</code> and <code>test_differential_abundance</code>.</p>
</div>
<div id="scaling-counts-to-normalise" class="section level1">
<h1 class="hasAnchor">
<a href="#scaling-counts-to-normalise" class="anchor"></a>Scaling counts to normalise</h1>
<p>Scaling of counts, normalisation, is performed to eliminate uninteresting differences between samples. These are differences due to sequencing depth or composition. In the tidybulk package the function <code>scale_abundance</code> generates scaled counts, it will also perform the filtering. We can choose from different normalisation methods. Here we will use the default, TMM, <span class="citation">[@robinson2010tmm]</span>. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co"># Scaling counts</span>
<span class="no">counts_scaled</span> <span class="kw">&lt;-</span> <span class="no">counts_tt</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span>(<span class="kw">factor_of_interest</span> <span class="kw">=</span> <span class="no">dex</span>)

<span class="co"># take a look</span>
<span class="no">counts_scaled</span>
<span class="co">#&gt; # A tibble: 512,816 x 18</span>
<span class="co">#&gt;    feature sample counts SampleName cell  dex   albut Run   avgLength Experiment</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;     </span>
<span class="co">#&gt;  1 ENSG00… 508       679 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  2 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  3 ENSG00… 508       467 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  4 ENSG00… 508       260 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  5 ENSG00… 508        60 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  6 ENSG00… 508         0 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  7 ENSG00… 508      3251 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  8 ENSG00… 508      1433 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt;  9 ENSG00… 508       519 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; 10 ENSG00… 508       394 GSM1275862 N613… untrt untrt SRR1…       126 SRX384345 </span>
<span class="co">#&gt; # … with 512,806 more rows, and 8 more variables: Sample &lt;fct&gt;,</span>
<span class="co">#&gt; #   BioSample &lt;fct&gt;, transcript &lt;chr&gt;, ref_genome &lt;chr&gt;, counts_scaled &lt;dbl&gt;,</span>
<span class="co">#&gt; #   TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;, lowly_abundant &lt;lgl&gt;</span></pre></body></html></div>
<p>After we run <code>scale_abundance</code> we should see some columns have been added at the end. We have a column called <code>lowly_abundant</code> that indicates whether the gene has been filtered due to being lowly expressed. FALSE means the gene wasn’t filtered, TRUE means it was. The <code>counts_scaled</code> column contains the scaled counts.</p>
<p>We can now see the difference of abundance densities before and after scaling.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="kw">cols</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"counts"</span>, <span class="st">"counts_scaled"</span>), <span class="kw">names_to</span> <span class="kw">=</span> <span class="st">"source"</span>, <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"abundance"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">abundance</span> + <span class="fl">1</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">color</span><span class="kw">=</span><span class="no">sample</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">source</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div id="exercises" class="section level4">
<h4 class="hasAnchor">
<a href="#exercises" class="anchor"></a>Exercises</h4>
<ol style="list-style-type: decimal">
<li>To see the effect of the filtering, adapt the code above to create density plots of the genes without filtering lowly expressed. How does the with and without filtering compare?</li>
<li>Another way to check the distributions of the counts in the samples is with box plots. Adapt the code above to create box plots of the counts before and after scaling. How does the before and after compare?</li>
</ol>
<p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from the paper (here)[<a href="https://f1000research.com/articles/5-1408" class="uri">https://f1000research.com/articles/5-1408</a>]).</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># tidybulk</span>
<span class="no">airway</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span>(<span class="kw">cols</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"counts"</span>, <span class="st">"counts_scaled"</span>), <span class="kw">names_to</span> <span class="kw">=</span> <span class="st">"source"</span>, <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"abundance"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">abundance</span> + <span class="fl">1</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">sample</span>, <span class="kw">color</span><span class="kw">=</span><span class="no">sample</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">source</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()

<span class="co"># standard method using edgeR</span>
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="fu">SE2DGEList</span>(<span class="no">airway</span>)
<span class="no">keep.exprs</span> <span class="kw">&lt;-</span> <span class="fu">filterByExpr</span>(<span class="no">dgList</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">dgList</span>$<span class="no">samples</span>$<span class="no">dex</span>)
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="no">dgList</span>[<span class="no">keep.exprs</span>,, <span class="kw">keep.lib.sizes</span><span class="kw">=</span><span class="fl">FALSE</span>]
<span class="no">nsamples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">dgList</span>)
<span class="no">col</span> <span class="kw">&lt;-</span> <span class="kw pkg">RColorBrewer</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="no">nsamples</span>, <span class="st">"Paired"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mfrow</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">2</span>))
<span class="no">logcounts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="no">dgList</span>$<span class="no">counts</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span>(<span class="no">logcounts</span>[,<span class="fl">1</span>]), <span class="kw">col</span><span class="kw">=</span><span class="no">col</span>[<span class="fl">1</span>], <span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">ylim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">0.26</span>), <span class="kw">las</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">""</span>, <span class="kw">xlab</span><span class="kw">=</span><span class="st">""</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span>(<span class="kw">main</span><span class="kw">=</span><span class="st">"Counts"</span>)
<span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">2</span>:<span class="no">nsamples</span>){
  <span class="no">den</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span>(<span class="no">logcounts</span>[,<span class="no">i</span>])
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span>(<span class="no">den</span>$<span class="no">x</span>, <span class="no">den</span>$<span class="no">y</span>, <span class="kw">col</span><span class="kw">=</span><span class="no">col</span>[<span class="no">i</span>], <span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>)
}
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topright"</span>, <span class="kw">legend</span><span class="kw">=</span><span class="no">dgList</span>$<span class="no">samples</span>$<span class="no">Run</span>, <span class="kw">text.col</span><span class="kw">=</span><span class="no">col</span>, <span class="kw">bty</span><span class="kw">=</span><span class="st">"n"</span>)
<span class="no">dgList_norm</span> <span class="kw">&lt;-</span> <span class="fu">calcNormFactors</span>(<span class="no">dgList</span>)
<span class="no">lcpm_n</span> <span class="kw">&lt;-</span> <span class="fu">cpm</span>(<span class="no">dgList_norm</span>, <span class="kw">log</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span>(<span class="no">lcpm_n</span>[,<span class="fl">1</span>]), <span class="kw">col</span><span class="kw">=</span><span class="no">col</span>[<span class="fl">1</span>], <span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">ylim</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">0.26</span>), <span class="kw">las</span><span class="kw">=</span><span class="fl">2</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">""</span>, <span class="kw">xlab</span><span class="kw">=</span><span class="st">""</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span>(<span class="st">"Counts scaled"</span>)
<span class="kw">for</span> (<span class="no">i</span> <span class="kw">in</span> <span class="fl">2</span>:<span class="no">nsamples</span>){
  <span class="no">den</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/density.html">density</a></span>(<span class="no">lcpm_n</span>[,<span class="no">i</span>])
  <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html">lines</a></span>(<span class="no">den</span>$<span class="no">x</span>, <span class="no">den</span>$<span class="no">y</span>, <span class="kw">col</span><span class="kw">=</span><span class="no">col</span>[<span class="no">i</span>], <span class="kw">lwd</span><span class="kw">=</span><span class="fl">2</span>)
}
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topright"</span>, <span class="kw">legend</span><span class="kw">=</span><span class="no">dgList_norm</span>$<span class="no">samples</span>$<span class="no">Run</span>, <span class="kw">text.col</span><span class="kw">=</span><span class="no">col</span>, <span class="kw">bty</span><span class="kw">=</span><span class="st">"n"</span>)</pre></body></html></div>
</div>
</div>
<div id="quality-control" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h1>
<div id="multidimensional-scaling-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#multidimensional-scaling-plots" class="anchor"></a>Multidimensional scaling plots</h2>
<p>By far, one of the most important plots we make when we analyse RNA-Seq data are MDS plots. An MDS plot is a visualisation of a principal components analysis, which determines the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. We can use the <code>reduce_dimensions</code> function to calculate the dimensions.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># get MDS dimensions</span>
<span class="no">counts_scal_MDS</span> <span class="kw">&lt;-</span>
  <span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span>(<span class="kw">method</span><span class="kw">=</span><span class="st">"MDS"</span>, <span class="kw">.dims</span> <span class="kw">=</span> <span class="fl">2</span>)
<span class="co">#&gt; tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</span></pre></body></html></div>
<p>This joins the result to the counts object. For plotting, we can select just the dimensions for the samples with <code>pivot_sample</code>.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># take a look</span>
<span class="no">counts_scal_MDS</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span>()
<span class="co">#&gt; # A tibble: 8 x 14</span>
<span class="co">#&gt;   sample SampleName cell  dex   albut Run   avgLength Experiment Sample</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;fct&gt;      &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;     &lt;int&gt; &lt;fct&gt;      &lt;fct&gt; </span>
<span class="co">#&gt; 1 508    GSM1275862 N613… untrt untrt SRR1…       126 SRX384345  SRS50…</span>
<span class="co">#&gt; 2 509    GSM1275863 N613… trt   untrt SRR1…       126 SRX384346  SRS50…</span>
<span class="co">#&gt; 3 512    GSM1275866 N052… untrt untrt SRR1…       126 SRX384349  SRS50…</span>
<span class="co">#&gt; 4 513    GSM1275867 N052… trt   untrt SRR1…        87 SRX384350  SRS50…</span>
<span class="co">#&gt; 5 516    GSM1275870 N080… untrt untrt SRR1…       120 SRX384353  SRS50…</span>
<span class="co">#&gt; 6 517    GSM1275871 N080… trt   untrt SRR1…       126 SRX384354  SRS50…</span>
<span class="co">#&gt; 7 520    GSM1275874 N061… untrt untrt SRR1…       101 SRX384357  SRS50…</span>
<span class="co">#&gt; 8 521    GSM1275875 N061… trt   untrt SRR1…        98 SRX384358  SRS50…</span>
<span class="co">#&gt; # … with 5 more variables: BioSample &lt;fct&gt;, TMM &lt;dbl&gt;, multiplier &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Dim1 &lt;dbl&gt;, Dim2 &lt;dbl&gt;</span></pre></body></html></div>
<p>We can now plot the reduced dimensions</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># MDS plot</span>
<span class="no">counts_scal_MDS</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">Dim1</span>, <span class="kw">y</span><span class="kw">=</span><span class="no">Dim2</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">dex</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div id="exercises-1" class="section level4">
<h4 class="hasAnchor">
<a href="#exercises-1" class="anchor"></a>Exercises</h4>
<ol style="list-style-type: decimal">
<li>Colour the MDS plot by cell line. Try using <code>shape</code> inside <code>aes</code> with the dex or cell variables.</li>
<li>Discuss what is the greatest source of variation in the data (i.e. what does dimension 1 represent)?</li>
</ol>
</div>
</div>
<div id="hierarchical-clustering-with-heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#hierarchical-clustering-with-heatmaps" class="anchor"></a>Hierarchical clustering with heatmaps</h2>
<p>An alternative to MDS for examining relationships between samples is using hierarchical clustering. Heatmaps are a nice visualisation to examine hierarchical clustering of your samples. tidybulk has a simple function we can use, <code>keep_variable</code> to extract the most variable genes which we can then plot with tidyHeatmap.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>

    <span class="co"># extract 500 most variable genes</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/keep_variable-methods.html">keep_variable</a></span>( <span class="kw">.abundance</span> <span class="kw">=</span> <span class="no">counts_scaled</span>, <span class="kw">top</span> <span class="kw">=</span> <span class="fl">500</span>) <span class="kw">%&gt;%</span>

    <span class="co"># create heatmap</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidyHeatmap/man/heatmap.html">heatmap</a></span>(
          <span class="kw">.column</span> <span class="kw">=</span> <span class="no">sample</span>,
          <span class="kw">.row</span> <span class="kw">=</span> <span class="no">feature</span>,
          <span class="kw">.value</span> <span class="kw">=</span> <span class="no">counts_scaled</span>,
          <span class="kw">annotation</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">dex</span>, <span class="no">cell</span>),
          <span class="kw">transform</span> <span class="kw">=</span> <span class="no">log1p</span>
      )
<span class="co">#&gt; Getting the 500 most variable genes</span></pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-14-1.png" width="768"></p>
</div>
</div>
<div id="differential-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h1>
<p>Now that we are happy that the data looks good, we can continue to testing for differentially expressed genes. We will use the <code>test_differential_abundance</code> from tidybulk which currently uses <em>edgeR</em> to perform the differential expression analysis. We give <code>test_differential_abundance</code> our tidybulk counts object and a formula, specifying the column that contains our groups to be compared. If all our samples were from the same cell line we could use the formula <code>0 + dex</code>, however, each treated and untreated sample is from a different cell line so we add the cell line as an additional factor <code>0 + dex + cell</code>. We also provide the names of the groups we want to compare to .contrasts (e.g. .contrasts = c(“dextreat - dexuntreat”)). <code>test_differential_abundance</code> will perform the filtering of lowly expressed genes as described before. The results will be joined to our counts for every sample. If we just want a table of differentially expressed genes we can use <code>pivot_transcript</code>.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">counts_de</span> <span class="kw">&lt;-</span> <span class="no">counts_tt</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span>(
      <span class="kw">.formula</span> <span class="kw">=</span> ~ <span class="fl">0</span> + <span class="no">dex</span> + <span class="no">cell</span>,
      <span class="kw">.contrasts</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dextrt - dexuntrt"</span>))
<span class="co">#&gt; tidybulk says: The design column names are "dextrt, dexuntrt, cellN061011, cellN080611, cellN61311"</span>
<span class="co">#&gt; tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$edgeR`</span>

<span class="co"># take a look</span>
<span class="no">counts_de</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>()
<span class="co">#&gt; # A tibble: 64,102 x 11</span>
<span class="co">#&gt;    feature albut transcript ref_genome `logFC_dextrt -… `logCPM_dextrt …</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;fct&gt; &lt;chr&gt;      &lt;chr&gt;                 &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG00… untrt TSPAN6     hg38                -0.390              5.06</span>
<span class="co">#&gt;  2 ENSG00… untrt TNMD       hg38                NA                 NA   </span>
<span class="co">#&gt;  3 ENSG00… untrt DPM1       hg38                 0.198              4.61</span>
<span class="co">#&gt;  4 ENSG00… untrt SCYL3      hg38                 0.0292             3.48</span>
<span class="co">#&gt;  5 ENSG00… untrt C1orf112   hg38                -0.124              1.47</span>
<span class="co">#&gt;  6 ENSG00… untrt FGR        hg38                NA                 NA   </span>
<span class="co">#&gt;  7 ENSG00… untrt CFH        hg38                 0.417              8.09</span>
<span class="co">#&gt;  8 ENSG00… untrt FUCA2      hg38                -0.250              5.91</span>
<span class="co">#&gt;  9 ENSG00… untrt GCLC       hg38                -0.0581             4.84</span>
<span class="co">#&gt; 10 ENSG00… untrt NFYA       hg38                -0.509              4.13</span>
<span class="co">#&gt; # … with 64,092 more rows, and 5 more variables: `F_dextrt - dexuntrt` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `PValue_dextrt - dexuntrt` &lt;dbl&gt;, `FDR_dextrt - dexuntrt` &lt;dbl&gt;,</span>
<span class="co">#&gt; #   `significant_dextrt - dexuntrt` &lt;lgl&gt;, lowly_abundant &lt;lgl&gt;</span></pre></body></html></div>
<p>Now we have columns with our logFC and FDR P values. There is also a column called significant that indicates if the gene has FDR &lt; 0.05.</p>
<p>Tidybulk enables a simplified way of performing an RNA-seq differential expression testing (with the benefit of smoothly integrating with ggplot2 and other tidyverse packages). Compare the code for a tidybulk edgeR analysis versus standard edgeR below.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># tidybulk with edgeR</span>
<span class="no">counts_m</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span>(<span class="kw">.formula</span> <span class="kw">=</span> ~ <span class="fl">0</span> + <span class="no">dex</span> + <span class="no">cell</span>, <span class="kw">.contrasts</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dextrt - dexuntrt"</span>))

<span class="co"># standard edgeR</span>
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="fu">DGEList</span>(<span class="kw">counts</span><span class="kw">=</span><span class="no">counts_m</span>, <span class="kw">group</span><span class="kw">=</span><span class="no">dex</span>)
<span class="no">keep</span> <span class="kw">&lt;-</span> <span class="fu">filterByExpr</span>(<span class="no">dgList</span>)
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="no">dgList</span>[<span class="no">keep</span>,,<span class="kw">keep.lib.sizes</span><span class="kw">=</span><span class="fl">FALSE</span>]
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="fu">calcNormFactors</span>(<span class="no">dgList</span>)
<span class="no">design</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span>(~ <span class="fl">0</span> + <span class="no">dex</span> + <span class="no">cell</span>)
<span class="no">dgList</span> <span class="kw">&lt;-</span> <span class="fu">estimateDisp</span>(<span class="no">dgList</span>,<span class="no">design</span>)
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu">glmQLFit</span>(<span class="no">dgList</span>,<span class="no">design</span>)
<span class="no">qlf</span> <span class="kw">&lt;-</span> <span class="fu">glmQLFTest</span>(<span class="no">fit</span>,<span class="kw">coef</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<div id="table-of-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#table-of-differentially-expressed-genes" class="anchor"></a>Table of differentially expressed genes</h3>
<p>We can write out our differentially expressed genes to a file that can be loaded into e.g. Excel. <code>write_tsv</code> will create a tab-separated file.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># save results</span>
<span class="no">counts_de</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span>(<span class="st">"de_results.tsv"</span>)</pre></body></html></div>
</div>
<div id="counting-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#counting-differentially-expressed-genes" class="anchor"></a>Counting differentially expressed genes</h3>
<p>We only have one contrast here so we could remove the suffix <code>_dextrt - dexuntrt</code> from the column headers to make the names shorter to use in the rest of the commands here, using <code>rename_at</code> package together with <code>str_replace</code>.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">&lt;-</span>
    <span class="no">counts_de</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_at</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/vars.html">vars</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">contains</a></span>(<span class="st">"dex"</span>)),
            ~<span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span>(<span class="no">.</span>, <span class="st">"_dextrt - dexuntrt"</span>, <span class="st">""</span>))</pre></body></html></div>
<p>We can count how many differentially expressed genes there are using the significant column.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">significant</span> <span class="kw">==</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span>(<span class="kw">num_de</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html">n_distinct</a></span>(<span class="no">transcript</span>))
<span class="co">#&gt; # A tibble: 1 x 1</span>
<span class="co">#&gt;   num_de</span>
<span class="co">#&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   4739</span></pre></body></html></div>
<div id="exercise" class="section level4">
<h4 class="hasAnchor">
<a href="#exercise" class="anchor"></a>Exercise</h4>
<p>Count how many upregulated DE genes there are.<br>
Count how many downregulated DE genes there are.</p>
</div>
</div>
<div id="extracting-top-differentially-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#extracting-top-differentially-expressed-genes" class="anchor"></a>Extracting top differentially expressed genes</h3>
<p>We can see the top genes by smallest P value. We’ll take a look at the top 6.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">topgenes</span> <span class="kw">&lt;-</span> <span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="no">PValue</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fl">6</span>)

<span class="no">topgenes</span>
<span class="co">#&gt; # A tibble: 6 x 11</span>
<span class="co">#&gt;   feature albut transcript ref_genome logFC logCPM     F   PValue     FDR</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;fct&gt; &lt;chr&gt;      &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00… untrt CACNB2     hg38        3.28   4.51 1575. 3.34e-11 4.07e-7</span>
<span class="co">#&gt; 2 ENSG00… untrt ZBTB16     hg38        7.15   4.15 1429. 5.11e-11 4.07e-7</span>
<span class="co">#&gt; 3 ENSG00… untrt PRSS35     hg38       -2.76   3.91  807. 6.16e-10 2.57e-6</span>
<span class="co">#&gt; 4 ENSG00… untrt PDPN       hg38        1.88   5.68  768. 7.60e-10 2.57e-6</span>
<span class="co">#&gt; 5 ENSG00… untrt SPARCL1    hg38        4.56   5.53  721. 1.00e- 9 2.57e-6</span>
<span class="co">#&gt; 6 ENSG00… untrt DUSP1      hg38        2.94   7.31  694. 1.18e- 9 2.57e-6</span>
<span class="co">#&gt; # … with 2 more variables: significant &lt;lgl&gt;, lowly_abundant &lt;lgl&gt;</span></pre></body></html></div>
<p>We can extract the symbols for these top genes to use in some of the plots we will make.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">topgenes_symbols</span> <span class="kw">&lt;-</span> <span class="no">topgenes</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span>(<span class="no">transcript</span>)

<span class="co"># take a look</span>
<span class="no">topgenes_symbols</span>
<span class="co">#&gt; [1] "CACNB2"  "ZBTB16"  "PRSS35"  "PDPN"    "SPARCL1" "DUSP1"</span></pre></body></html></div>
</div>
</div>
<div id="plots-after-testing-for-de" class="section level1">
<h1 class="hasAnchor">
<a href="#plots-after-testing-for-de" class="anchor"></a>Plots after testing for DE</h1>
<p>Let’s make a few plots to make sure everything looks good and that we haven’t made a mistake in the analysis. Genome-wide plots that are useful for checking are MA plots and volcano plots. We can also use stripcharts to visualise groups of genes.</p>
<div id="ma-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#ma-plots" class="anchor"></a>MA plots</h3>
<p>MA plots enable us to visualise <strong>amount</strong> of expression (logCPM) versus logFC. Highly expressed genes are towards the right of the plot. We can also colour significant genes (e.g. genes with FDR &lt; 0.05)</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="co"># maplot, minimal</span>
<span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">logCPM</span>, <span class="kw">y</span><span class="kw">=</span>-<span class="no">logFC</span>, <span class="kw">colour</span><span class="kw">=</span><span class="no">significant</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>A more informative plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>

    <span class="co"># Subset data</span>
        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">significant</span> <span class="kw">=</span> <span class="no">FDR</span><span class="kw">&lt;</span><span class="fl">0.05</span> <span class="kw">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;=</span><span class="fl">2</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">transcript</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;=</span><span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">transcript</span>), <span class="st">""</span>)) <span class="kw">%&gt;%</span>

    <span class="co"># Plot</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">logCPM</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">logFC</span>, <span class="kw">label</span><span class="kw">=</span><span class="no">transcript</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">size</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="no">significant</span>)) +
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"#e11f28"</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span>(<span class="kw">range</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">2</span>)) +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div id="volcano-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h3>
<p>Volcano plots enable us to visualise <strong>significance</strong> of expression (P value) versus logFC. Highly significant genes are towards the top of the plot. We can also colour significant genes (e.g. genes with FDR &lt; 0.05)</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co"># volcanoplot, minimal</span>
<span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span><span class="kw">=</span><span class="no">logFC</span>, <span class="kw">y</span><span class="kw">=</span>-<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">PValue</span>), <span class="kw">colour</span><span class="kw">=</span><span class="no">significant</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>A more informative plot, integrating some of the packages in tidyverse.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">counts_de_pretty</span> <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_transcript-methods.html">pivot_transcript</a></span>() <span class="kw">%&gt;%</span>

    <span class="co"># Subset data</span>
        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(!<span class="no">lowly_abundant</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">significant</span> <span class="kw">=</span> <span class="no">FDR</span><span class="kw">&lt;</span><span class="fl">0.05</span> <span class="kw">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">logFC</span>) <span class="kw">&gt;=</span><span class="fl">2</span>) <span class="kw">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">transcript</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">transcript</span> <span class="kw">%in%</span> <span class="no">topgenes_symbols</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">transcript</span>), <span class="st">""</span>)) <span class="kw">%&gt;%</span>

    <span class="co"># Plot</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">logFC</span>, <span class="kw">y</span> <span class="kw">=</span> -<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="no">PValue</span>), <span class="kw">label</span><span class="kw">=</span><span class="no">transcript</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">size</span> <span class="kw">=</span> <span class="no">significant</span>, <span class="kw">alpha</span><span class="kw">=</span><span class="no">significant</span>)) +
    <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"#e11f28"</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span>(<span class="kw">range</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">2</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">10</span>,<span class="fl">10</span>)) +
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div id="stripcharts" class="section level3">
<h3 class="hasAnchor">
<a href="#stripcharts" class="anchor"></a>Stripcharts</h3>
<p>In addition to the genome-wide plots already discussed, it is recommended to have a look at the expression levels of the individual samples for the genes of interest, before following up on the DE genes with further lab work. We can use stripcharts to do this. These will help show if expression is consistent amongst replicates in the groups.</p>
<p>With stripcharts we can see if replicates tend to group together and how the expression compares to the other groups. We’ll also add a box plot to show the distribution.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>

    <span class="co"># extract counts for top DE genes</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">transcript</span> <span class="kw">%in%</span> <span class="no">topgenes_symbols</span>) <span class="kw">%&gt;%</span>

    <span class="co"># make stripchart</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">dex</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">counts_scaled</span> + <span class="fl">1</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">dex</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">transcript</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="tidytranscriptomics_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div id="interactive-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#interactive-plots" class="anchor"></a>Interactive Plots</h3>
<p>A really nice feature of using tidyverse and ggplot2 is that we can make interactive plots quite easily using the plotly package. This can be very useful for exploring what genes or samples are in the plots. To make interactive plots we first save our ggplot2 object, for example as “p”. We can add label= into the <code>aes</code> if we want other variables, such as the sample name to be visible when we hover over the plot.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">p</span> <span class="kw">&lt;-</span> <span class="no">counts_scaled</span> <span class="kw">%&gt;%</span>

    <span class="co"># extract counts for top DE genes</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">transcript</span> <span class="kw">%in%</span> <span class="no">topgenes_symbols</span>) <span class="kw">%&gt;%</span>

  <span class="co"># make stripchart</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">dex</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">counts_scaled</span> + <span class="fl">1</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">dex</span>, <span class="kw">label</span> <span class="kw">=</span> <span class="no">sample</span>)) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>() +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">transcript</span>) +
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>()+
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p>Then we simply give our plot, p, to the ggplotly function. We can also specify which parameters from the aes() we want to show up when we hover over the plot with <code>tooltip</code>.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span>(<span class="no">p</span>, <span class="kw">tooltip</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"label"</span>, <span class="st">"y"</span>))
<span class="co">#&gt; Warning: `group_by_()` is deprecated as of dplyr 0.7.0.</span>
<span class="co">#&gt; Please use `group_by()` instead.</span>
<span class="co">#&gt; See vignette('programming') for more help</span>
<span class="co">#&gt; This warning is displayed once every 8 hours.</span>
<span class="co">#&gt; Call `lifecycle::last_warnings()` to see where this warning was generated.</span></pre></body></html></div>
<div id="htmlwidget-be1fb91a55ffd52c14fe" style="width:700px;height:432.632880098888px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-be1fb91a55ffd52c14fe">{"x":{"data":[{"x":[1,1,1,1],"y":[2.69191110980537,3.07411718730643,2.89495322609288,3.00394565317732],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.73126672017509,3.81662547175731,3.63836213210247,3.84475635521907],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x2","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[3.20889429106159,3.48668622056427,2.82592920686795,3.12089329404981],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x3","yaxis":"y","frame":null},{"x":[1,1,1,1],"y":[1.65303020004975,2.11216518033722,1.93173855829516,1.79481170281122],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[3.34031094320166,3.2285919684294,3.12125503477529,3.33984448642205],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[1,1,1,1],"y":[2.89967429955968,2.79256577041858,2.6019847639754,3.05651479236326],"hoverinfo":"y","type":"box","fillcolor":"rgba(248,118,109,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[1.88512386656772,2.08729004672719,1.75408369787238,2.00793353169738],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":true,"xaxis":"x","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.79984947368254,3.00070054302364,2.86395616346094,2.82619456899666],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x2","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.9267308485033,2.62330477835599,2.25075450230717,2.57641323591571],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x3","yaxis":"y","frame":null},{"x":[2,2,2,2],"y":[2.4962401879048,2.87459328894999,2.77192330101204,2.66478216152997],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[1.77606510611906,1.91732676692718,1.91911963767485,1.92420606242723],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x2","yaxis":"y2","frame":null},{"x":[2,2,2,2],"y":[0.680145244542111,1.1337127133929,0.26009033496043,0.709655221922282],"hoverinfo":"y","type":"box","fillcolor":"rgba(0,191,196,1)","marker":{"opacity":null,"outliercolor":"rgba(0,0,0,1)","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"},"size":5.66929133858268},"line":{"color":"rgba(51,51,51,1)","width":1.88976377952756},"name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x3","yaxis":"y2","frame":null},{"x":[1.02976017091423,1.1011077772826,0.910661433078349,0.750850533507764],"y":[2.8948607296932,3.07404651137892,2.69198383578681,3.00404178042232],"text":["counts_scaled + 1:  785.151069<br />sample: 509","counts_scaled + 1: 1186.088752<br />sample: 513","counts_scaled + 1:  491.938837<br />sample: 517","counts_scaled + 1: 1009.126598<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.23512257486582,1.08578401785344,1.22428520657122,1.14076947793365],"y":[3.73130355060584,3.81659786336053,3.63828476649025,3.84469622736453],"text":["counts_scaled + 1: 5386.004602<br />sample: 509","counts_scaled + 1: 6555.796616<br />sample: 513","counts_scaled + 1: 4348.726873<br />sample: 517","counts_scaled + 1: 6994.494858<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.31202957369387,0.772038727998734,1.28061306290329,0.944536003656685],"y":[3.48660564664029,3.20893655517353,2.82587370187896,3.12078787949134],"text":["counts_scaled + 1: 3066.805409<br />sample: 509","counts_scaled + 1: 1617.686239<br />sample: 513","counts_scaled + 1:  669.775422<br />sample: 517","counts_scaled + 1: 1320.971032<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0.840161112137139,0.917321312241256,0.837581389211118,1.202171706222],"y":[1.65294844883136,2.11213410179811,1.93178338073078,1.79475212807491],"text":["counts_scaled + 1:   44.981113<br />sample: 509","counts_scaled + 1:  129.468817<br />sample: 513","counts_scaled + 1:   85.455212<br />sample: 517","counts_scaled + 1:   62.346446<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.23863838165998,0.868008347414434,1.25399169307202,0.646496980637312],"y":[3.34029870085944,3.22853029235943,3.12131085769151,3.33989153502654],"text":["counts_scaled + 1: 2189.328563<br />sample: 509","counts_scaled + 1: 1692.746671<br />sample: 513","counts_scaled + 1: 1322.071779<br />sample: 517","counts_scaled + 1: 2186.978363<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.16592773087323,0.991402395069599,1.32322251331061,1.37806720472872],"y":[2.89965708194095,2.79259539696262,2.60190789231367,3.05651294419462],"text":["counts_scaled + 1:  793.732749<br />sample: 509","counts_scaled + 1:  620.248568<br />sample: 513","counts_scaled + 1:  399.930719<br />sample: 517","counts_scaled + 1: 1138.976576<br />sample: 521"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(248,118,109,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"trt","legendgroup":"trt","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[2.16029344629496,1.80662347637117,1.93565584588796,2.15369586274028],"y":[1.88503973428959,2.08727475910885,1.75410989438484,2.00786707896551],"text":["counts_scaled + 1:   76.758038<br />sample: 508","counts_scaled + 1:  122.261592<br />sample: 512","counts_scaled + 1:   56.765399<br />sample: 516","counts_scaled + 1:  101.843551<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.81010362580419,2.09480238519609,1.82777483630925,1.98742575272918],"y":[2.79988554518442,3.00074079816815,2.82624344134077,2.86399337245268],"text":["counts_scaled + 1:  630.738693<br />sample: 508","counts_scaled + 1: 1001.614362<br />sample: 512","counts_scaled + 1:  670.184793<br />sample: 516","counts_scaled + 1:  731.065288<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x2","yaxis":"y","hoverinfo":"text","frame":null},{"x":[2.35073908828199,1.95622447934002,1.953322776407,1.89633470419794],"y":[2.92667158879708,2.62335383895801,2.25079989418616,2.57636821502637],"text":["counts_scaled + 1:  844.755151<br />sample: 508","counts_scaled + 1:  420.053665<br />sample: 512","counts_scaled + 1:  178.137151<br />sample: 516","counts_scaled + 1:  377.062407<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x3","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1.63978125397116,1.97507381420583,2.36856076996773,1.86194462943822],"y":[2.49629435839199,2.87460310210155,2.77199261771356,2.66468587650072],"text":["counts_scaled + 1:  313.501908<br />sample: 508","counts_scaled + 1:  749.192272<br />sample: 512","counts_scaled + 1:  591.457170<br />sample: 516","counts_scaled + 1:  462.149153<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[1.66871780604124,1.80714368168265,1.64205385949463,2.18225637692958],"y":[1.77609837853409,1.91741281627406,1.91921682501229,1.92420527216412],"text":["counts_scaled + 1:   59.712480<br />sample: 508","counts_scaled + 1:   82.665970<br />sample: 512","counts_scaled + 1:   83.007940<br />sample: 516","counts_scaled + 1:   83.985838<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x2","yaxis":"y2","hoverinfo":"text","frame":null},{"x":[2.31506570614874,2.24557304289192,1.87151676490903,2.21435323413461],"y":[0.680172027716959,0.709656039835411,0.260116047514121,1.13375763602523],"text":["counts_scaled + 1:    4.787902<br />sample: 508","counts_scaled + 1:    5.124544<br />sample: 512","counts_scaled + 1:    1.820079<br />sample: 516","counts_scaled + 1:   13.605444<br />sample: 520"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,191,196,1)","opacity":1,"size":5.66929133858268,"symbol":"circle","line":{"width":1.88976377952756,"color":"rgba(0,0,0,1)"}},"hoveron":"points","name":"untrt","legendgroup":"untrt","showlegend":false,"xaxis":"x3","yaxis":"y2","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":39.5141341357194,"r":7.30593607305936,"b":41.7789743183678,"l":54.7945205479452},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xaxis":{"domain":[0,0.32941943900848],"automargin":true,"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"annotations":[{"text":"dex","x":0.5,"y":-0.0196373426241191,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"top","annotationType":"axis"},{"text":"counts_scaled + 1","x":-0.0215264187866928,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis"},{"text":"CACNB2","x":0.16470971950424,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"DUSP1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PDPN","x":0.83529028049576,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"PRSS35","x":0.16470971950424,"y":0.483529970702352,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"SPARCL1","x":0.5,"y":0.483529970702352,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"ZBTB16","x":0.83529028049576,"y":0.483529970702352,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(26,26,26,1)","family":"","size":11.689497716895},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"},{"text":"dex","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"yaxis":{"domain":[0.516470029297648,1],"automargin":true,"type":"linear","autorange":false,"range":[0.0808570339474975,4.023989656232],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0.516470029297648,"y1":1},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":23.37899543379,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0.516470029297648,"y1":1},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":23.37899543379,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0.516470029297648,"y1":1},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":23.37899543379,"yanchor":1,"ysizemode":"pixel"},{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":0.483529970702352},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":0.32941943900848,"y0":0,"y1":23.37899543379,"yanchor":0.483529970702352,"ysizemode":"pixel"},{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":0.483529970702352},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.337247227658187,"x1":0.662752772341813,"y0":0,"y1":23.37899543379,"yanchor":0.483529970702352,"ysizemode":"pixel"},{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":0.483529970702352},{"type":"rect","fillcolor":"rgba(217,217,217,1)","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0.67058056099152,"x1":1,"y0":0,"y1":23.37899543379,"yanchor":0.483529970702352,"ysizemode":"pixel"}],"xaxis2":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"domain":[0.337247227658187,0.662752772341813],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"xaxis3":{"type":"linear","autorange":false,"range":[0.4,2.6],"tickmode":"array","ticktext":["trt","untrt"],"tickvals":[1,2],"categoryorder":"array","categoryarray":["trt","untrt"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"domain":[0.67058056099152,1],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y2","title":"","hoverformat":".2f"},"yaxis2":{"type":"linear","autorange":false,"range":[0.0808570339474975,4.023989656232],"tickmode":"array","ticktext":["10","100","1000","10000"],"tickvals":[1,2,3,4],"categoryorder":"array","categoryarray":["10","100","1000","10000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"domain":[0,0.483529970702352],"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.96395248281826},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"3c13b76b0c9":{"x":{},"y":{},"fill":{},"label":{},"type":"box"},"3c11a09e4f8":{"x":{},"y":{},"fill":{},"label":{}}},"cur_data":"3c13b76b0c9","visdat":{"3c13b76b0c9":["function (y) ","x"],"3c11a09e4f8":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>Record package and version information with <code>sessionInfo</code></p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.0 (2020-04-24)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt;  [1] parallel  stats4    grid      stats     graphics  grDevices utils    </span>
<span class="co">#&gt;  [8] datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] airway_1.9.0                SummarizedExperiment_1.19.5</span>
<span class="co">#&gt;  [3] DelayedArray_0.15.6         matrixStats_0.56.0         </span>
<span class="co">#&gt;  [5] Matrix_1.2-18               Biobase_2.49.0             </span>
<span class="co">#&gt;  [7] GenomicRanges_1.41.5        GenomeInfoDb_1.25.5        </span>
<span class="co">#&gt;  [9] IRanges_2.23.10             S4Vectors_0.27.12          </span>
<span class="co">#&gt; [11] BiocGenerics_0.35.4         plotly_4.9.2.1             </span>
<span class="co">#&gt; [13] ggrepel_0.8.2               tidybulk_1.1.3             </span>
<span class="co">#&gt; [15] tidyHeatmap_1.0.1           ComplexHeatmap_2.5.3       </span>
<span class="co">#&gt; [17] ggplot2_3.3.2               stringr_1.4.0              </span>
<span class="co">#&gt; [19] readr_1.3.1                 tidyr_1.1.0                </span>
<span class="co">#&gt; [21] dplyr_1.0.0                 tibble_3.0.1               </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] viridis_0.5.1          httr_1.4.1             edgeR_3.31.4          </span>
<span class="co">#&gt;  [4] splines_4.0.0          jsonlite_1.7.0         viridisLite_0.3.0     </span>
<span class="co">#&gt;  [7] assertthat_0.2.1       GenomeInfoDbData_1.2.3 yaml_2.2.1            </span>
<span class="co">#&gt; [10] lattice_0.20-41        pillar_1.4.4           backports_1.1.8       </span>
<span class="co">#&gt; [13] limma_3.45.7           glue_1.4.1             digest_0.6.25         </span>
<span class="co">#&gt; [16] RColorBrewer_1.1-2     XVector_0.29.3         colorspace_1.4-1      </span>
<span class="co">#&gt; [19] htmltools_0.5.0        preprocessCore_1.51.0  pkgconfig_2.0.3       </span>
<span class="co">#&gt; [22] GetoptLong_1.0.0       zlibbioc_1.35.0        purrr_0.3.4           </span>
<span class="co">#&gt; [25] scales_1.1.1           farver_2.0.3           generics_0.0.2        </span>
<span class="co">#&gt; [28] ellipsis_0.3.1         withr_2.2.0            lazyeval_0.2.2        </span>
<span class="co">#&gt; [31] cli_2.0.2              magrittr_1.5           crayon_1.3.4          </span>
<span class="co">#&gt; [34] memoise_1.1.0          evaluate_0.14          fansi_0.4.1           </span>
<span class="co">#&gt; [37] fs_1.4.1               MASS_7.3-51.5          tools_4.0.0           </span>
<span class="co">#&gt; [40] data.table_1.12.8      hms_0.5.3              GlobalOptions_0.1.2   </span>
<span class="co">#&gt; [43] lifecycle_0.2.0        locfit_1.5-9.4         munsell_0.5.0         </span>
<span class="co">#&gt; [46] cluster_2.1.0          compiler_4.0.0         pkgdown_1.5.1         </span>
<span class="co">#&gt; [49] rlang_0.4.6            RCurl_1.98-1.2         rjson_0.2.20          </span>
<span class="co">#&gt; [52] htmlwidgets_1.5.1      circlize_0.4.10        crosstalk_1.1.0.1     </span>
<span class="co">#&gt; [55] labeling_0.3           bitops_1.0-6           rmarkdown_2.3         </span>
<span class="co">#&gt; [58] gtable_0.3.0           R6_2.4.1               gridExtra_2.3         </span>
<span class="co">#&gt; [61] knitr_1.29             utf8_1.1.4             clue_0.3-57           </span>
<span class="co">#&gt; [64] rprojroot_1.3-2        shape_1.4.4            desc_1.2.0            </span>
<span class="co">#&gt; [67] stringi_1.4.6          Rcpp_1.0.4.6           vctrs_0.3.1           </span>
<span class="co">#&gt; [70] png_0.1-7              tidyselect_1.1.0       xfun_0.15</span></pre></body></html></div>
</div>
<div id="key-points" class="section level1">
<h1 class="hasAnchor">
<a href="#key-points" class="anchor"></a>Key Points</h1>
<ul>
<li>RNA-Seq data can be analysed in a ‘tidy’ way using the packages tidyverse, tidybulk and tidyHeatmap</li>
<li>Key steps in an RNA-Seq analysis are filtering lowly expressed genes, normalisation for sequencing depth and composition, and testing for differential expression</li>
<li>MDS plots are very important for examining the quality of the data</li>
<li>Other useful plots for assessing RNA-Seq data are bar plots, density plots, box plots, MA plots, volcano plots, stripcharts and heatmaps. These can all be made in a ‘tidy’ way.</li>
</ul>
</div>
<div id="further-reading" class="section level1">
<h1 class="hasAnchor">
<a href="#further-reading" class="anchor"></a>Further Reading</h1>
<p><a href="https://f1000research.com/articles/5-1408">RNA-Seq analysis is easy as 1-2-3 with limma, Glimma and edgeR</a></p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Peter Mac, <a href="mailto:maria.doyle@petermac.org" class="email">maria.doyle@petermac.org</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>WEHI, <a href="mailto:mangiola.s@wehi.edu.au" class="email">mangiola.s@wehi.edu.au</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Maria Doyle.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
